// @google-cloud/storage/build/cjs/src/index.js
"use strict";













var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Notification = exports.Iam = exports.HmacKey = exports.File = exports.Channel = exports.Bucket = exports.Storage = exports.RETRYABLE_ERR_FN_DEFAULT = exports.IdempotencyStrategy = exports.ApiError = void 0;

var index_js_1 = require("./nodejs-common/index.js");
Object.defineProperty(exports, "ApiError", { enumerable: true, get: function () { return index_js_1.ApiError; } });
var storage_js_1 = require("./storage.js");
Object.defineProperty(exports, "IdempotencyStrategy", { enumerable: true, get: function () { return storage_js_1.IdempotencyStrategy; } });
Object.defineProperty(exports, "RETRYABLE_ERR_FN_DEFAULT", { enumerable: true, get: function () { return storage_js_1.RETRYABLE_ERR_FN_DEFAULT; } });
Object.defineProperty(exports, "Storage", { enumerable: true, get: function () { return storage_js_1.Storage; } });
var bucket_js_1 = require("./bucket.js");
Object.defineProperty(exports, "Bucket", { enumerable: true, get: function () { return bucket_js_1.Bucket; } });
__exportStar(require("./crc32c.js"), exports);
var channel_js_1 = require("./channel.js");
Object.defineProperty(exports, "Channel", { enumerable: true, get: function () { return channel_js_1.Channel; } });
var file_js_1 = require("./file.js");
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return file_js_1.File; } });
__exportStar(require("./hash-stream-validator.js"), exports);
var hmacKey_js_1 = require("./hmacKey.js");
Object.defineProperty(exports, "HmacKey", { enumerable: true, get: function () { return hmacKey_js_1.HmacKey; } });
var iam_js_1 = require("./iam.js");
Object.defineProperty(exports, "Iam", { enumerable: true, get: function () { return iam_js_1.Iam; } });
var notification_js_1 = require("./notification.js");
Object.defineProperty(exports, "Notification", { enumerable: true, get: function () { return notification_js_1.Notification; } });
__exportStar(require("./transfer-manager.js"), exports);
