// @google-cloud/storage/build/cjs/src/notification.js
"use strict";













Object.defineProperty(exports, "__esModule", { value: true });
exports.Notification = void 0;
const index_js_1 = require("./nodejs-common/index.js");
const promisify_1 = require("@google-cloud/promisify");


class Notification extends index_js_1.ServiceObject {
    constructor(bucket, id) {
        const requestQueryObject = {};
        const methods = {
            
            create: true,
            
            
            delete: {
                reqOpts: {
                    qs: requestQueryObject,
                },
            },
            
            get: {
                reqOpts: {
                    qs: requestQueryObject,
                },
            },
            
            getMetadata: {
                reqOpts: {
                    qs: requestQueryObject,
                },
            },
            
            
            
            exists: true,
        };
        super({
            parent: bucket,
            baseUrl: '/notificationConfigs',
            id: id.toString(),
            createMethod: bucket.createNotification.bind(bucket),
            methods,
        });
    }
}
exports.Notification = Notification;

(0, promisify_1.promisifyAll)(Notification);
