// @google-cloud/firestore/build/src/reference/composite-filter-internal.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.CompositeFilterInternal = void 0;
const filter_internal_1 = require("./filter-internal");
class CompositeFilterInternal extends filter_internal_1.FilterInternal {
    constructor(filters, operator) {
        super();
        this.filters = filters;
        this.operator = operator;


        this.memoizedFlattenedFilters = null;
    }
    getFilters() {
        return this.filters;
    }
    isConjunction() {
        return this.operator === 'AND';
    }
    getFlattenedFilters() {
        if (this.memoizedFlattenedFilters !== null) {
            return this.memoizedFlattenedFilters;
        }
        this.memoizedFlattenedFilters = this.filters.reduce((allFilters, subfilter) => allFilters.concat(subfilter.getFlattenedFilters()), []);
        return this.memoizedFlattenedFilters;
    }
    toProto() {
        if (this.filters.length === 1) {
            return this.filters[0].toProto();
        }
        const proto = {
            compositeFilter: {
                op: this.operator,
                filters: this.filters.map(filter => filter.toProto()),
            },
        };
        return proto;
    }
    isEqual(other) {
        if (other instanceof CompositeFilterInternal) {
            const otherFilters = other.getFilters();
            return (this.operator === other.operator &&
                this.getFilters().length === other.getFilters().length &&
                this.getFilters().every((filter, index) => filter.isEqual(otherFilters[index])));
        }
        else {
            return false;
        }
    }
}
exports.CompositeFilterInternal = CompositeFilterInternal;
