// @google-cloud/firestore/build/src/reference/field-filter-internal.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldFilterInternal = void 0;
const deepEqual = require("fast-deep-equal");
const filter_internal_1 = require("./filter-internal");

class FieldFilterInternal extends filter_internal_1.FilterInternal {
    getFlattenedFilters() {
        return [this];
    }
    getFilters() {
        return [this];
    }
    
    constructor(serializer, field, op, value) {
        super();
        this.serializer = serializer;
        this.field = field;
        this.op = op;
        this.value = value;
    }
    
    isInequalityFilter() {
        switch (this.op) {
            case 'GREATER_THAN':
            case 'GREATER_THAN_OR_EQUAL':
            case 'LESS_THAN':
            case 'LESS_THAN_OR_EQUAL':
            case 'NOT_EQUAL':
            case 'NOT_IN':
                return true;
            default:
                return false;
        }
    }
    
    toProto() {
        if (typeof this.value === 'number' && isNaN(this.value)) {
            return {
                unaryFilter: {
                    field: {
                        fieldPath: this.field.formattedName,
                    },
                    op: this.op === 'EQUAL' ? 'IS_NAN' : 'IS_NOT_NAN',
                },
            };
        }
        if (this.value === null) {
            return {
                unaryFilter: {
                    field: {
                        fieldPath: this.field.formattedName,
                    },
                    op: this.op === 'EQUAL' ? 'IS_NULL' : 'IS_NOT_NULL',
                },
            };
        }
        return {
            fieldFilter: {
                field: {
                    fieldPath: this.field.formattedName,
                },
                op: this.op,
                value: this.serializer.encodeValue(this.value),
            },
        };
    }
    isEqual(other) {
        return (other instanceof FieldFilterInternal &&
            this.field.isEqual(other.field) &&
            this.op === other.op &&
            deepEqual(this.value, other.value));
    }
}
exports.FieldFilterInternal = FieldFilterInternal;
