// @google-cloud/firestore/build/src/reference/query-snapshot.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.QuerySnapshot = void 0;
const validate_1 = require("../validate");
const util_1 = require("../util");

class QuerySnapshot {
    
    constructor(_query, _readTime, _size, docs, changes) {
        this._query = _query;
        this._readTime = _readTime;
        this._size = _size;
        this._materializedDocs = null;
        this._materializedChanges = null;
        this._docs = null;
        this._changes = null;
        this._docs = docs;
        this._changes = changes;
    }
    
    get query() {
        return this._query;
    }
    
    get docs() {
        if (this._materializedDocs) {
            return this._materializedDocs;
        }
        this._materializedDocs = this._docs();
        this._docs = null;
        return this._materializedDocs;
    }
    
    get empty() {
        return this._size === 0;
    }
    
    get size() {
        return this._size;
    }
    
    get readTime() {
        return this._readTime;
    }
    
    docChanges() {
        if (this._materializedChanges) {
            return this._materializedChanges;
        }
        this._materializedChanges = this._changes();
        this._changes = null;
        return this._materializedChanges;
    }
    
    forEach(callback, thisArg) {
        (0, validate_1.validateFunction)('callback', callback);
        for (const doc of this.docs) {
            callback.call(thisArg, doc);
        }
    }
    
    isEqual(other) {


        if (this === other) {
            return true;
        }
        if (!(other instanceof QuerySnapshot)) {
            return false;
        }
        if (this._size !== other._size) {
            return false;
        }
        if (!this._query.isEqual(other._query)) {
            return false;
        }
        if (this._materializedDocs && !this._materializedChanges) {

            return ((0, util_1.isArrayEqual)(this.docs, other.docs) &&
                (0, util_1.isArrayEqual)(this.docChanges(), other.docChanges()));
        }

        return ((0, util_1.isArrayEqual)(this.docChanges(), other.docChanges()) &&
            (0, util_1.isArrayEqual)(this.docs, other.docs));
    }
}
exports.QuerySnapshot = QuerySnapshot;
