// @google-cloud/firestore/build/src/reference/vector-query.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.VectorQuery = void 0;
const field_value_1 = require("../field-value");
const path_1 = require("../path");
const util_1 = require("../util");
const query_util_1 = require("./query-util");
const vector_query_snapshot_1 = require("./vector-query-snapshot");
const query_profile_1 = require("../query-profile");

class VectorQuery {
    
    constructor(_query, _options) {
        this._query = _query;
        this._options = _options;
        this._queryUtil = new query_util_1.QueryUtil(_query._firestore, _query._queryOptions, _query._serializer);
    }
    
    get query() {
        return this._query;
    }
    
    get _rawVectorField() {
        return typeof this._options.vectorField === 'string'
            ? this._options.vectorField
            : this._options.vectorField.toString();
    }
    
    get _rawDistanceResultField() {
        if (typeof this._options.distanceResultField === 'undefined')
            return;
        return typeof this._options.distanceResultField === 'string'
            ? this._options.distanceResultField
            : this._options.distanceResultField.toString();
    }
    
    get _rawQueryVector() {
        return Array.isArray(this._options.queryVector)
            ? this._options.queryVector
            : this._options.queryVector.toArray();
    }
    
    async explain(options) {
        if (options === undefined) {
            options = {};
        }
        const { result, explainMetrics } = await this._getResponse(options);
        if (!explainMetrics) {
            throw new Error('No explain results');
        }
        return new query_profile_1.ExplainResults(explainMetrics, result || null);
    }
    
    async get() {
        const { result } = await this._getResponse();
        if (!result) {
            throw new Error('No VectorQuerySnapshot result');
        }
        return result;
    }
    _getResponse(explainOptions) {
        return this._queryUtil._getResponse(this, 
         undefined, 

         false, explainOptions);
    }
    
    _stream(transactionId) {
        return this._queryUtil._stream(this, transactionId, 
         false);
    }
    
    toProto(transactionOrReadTime, explainOptions) {
        var _a, _b, _c;
        const queryProto = this._query.toProto(transactionOrReadTime);
        const queryVector = Array.isArray(this._options.queryVector)
            ? new field_value_1.VectorValue(this._options.queryVector)
            : this._options.queryVector;
        queryProto.structuredQuery.findNearest = {
            limit: { value: this._options.limit },
            distanceMeasure: this._options.distanceMeasure,
            vectorField: {
                fieldPath: path_1.FieldPath.fromArgument(this._options.vectorField)
                    .formattedName,
            },
            queryVector: queryVector._toProto(this._query._serializer),
            distanceResultField: ((_a = this._options) === null || _a === void 0 ? void 0 : _a.distanceResultField)
                ? path_1.FieldPath.fromArgument(this._options.distanceResultField)
                    .formattedName
                : undefined,
            distanceThreshold: ((_b = this._options) === null || _b === void 0 ? void 0 : _b.distanceThreshold)
                ? { value: (_c = this._options) === null || _c === void 0 ? void 0 : _c.distanceThreshold }
                : undefined,
        };
        if (explainOptions) {
            queryProto.explainOptions = explainOptions;
        }
        return queryProto;
    }
    
    _createSnapshot(readTime, size, docs, changes) {
        return new vector_query_snapshot_1.VectorQuerySnapshot(this, readTime, size, docs, changes);
    }
    
    startAfter(

    ...fieldValuesOrDocumentSnapshot) {
        throw new Error('Unimplemented: Vector query does not support cursors yet.');
    }
    
    isEqual(other) {
        if (this === other) {
            return true;
        }
        if (!(other instanceof VectorQuery)) {
            return false;
        }
        if (!this.query.isEqual(other.query)) {
            return false;
        }
        return (this._rawVectorField === other._rawVectorField &&
            (0, util_1.isPrimitiveArrayEqual)(this._rawQueryVector, other._rawQueryVector) &&
            this._options.limit === other._options.limit &&
            this._options.distanceMeasure === other._options.distanceMeasure &&
            this._options.distanceThreshold === other._options.distanceThreshold &&
            this._rawDistanceResultField === other._rawDistanceResultField);
    }
}
exports.VectorQuery = VectorQuery;
