// firebase-admin/lib/data-connect/data-connect.js

"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.DataConnect = exports.DataConnectService = void 0;
const data_connect_api_client_internal_1 = require("./data-connect-api-client-internal");
class DataConnectService {
    constructor(app) {
        this.dataConnectInstances = new Map();
        this.appInternal = app;
    }
    getDataConnect(connectorConfig) {
        const id = `${connectorConfig.location}-${connectorConfig.serviceId}`;
        const dc = this.dataConnectInstances.get(id);
        if (typeof dc !== 'undefined') {
            return dc;
        }
        const newInstance = new DataConnect(connectorConfig, this.appInternal);
        this.dataConnectInstances.set(id, newInstance);
        return newInstance;
    }
    
    get app() {
        return this.appInternal;
    }
}
exports.DataConnectService = DataConnectService;

class DataConnect {
    
    constructor(connectorConfig, app) {
        this.connectorConfig = connectorConfig;
        this.app = app;
        this.client = new data_connect_api_client_internal_1.DataConnectApiClient(connectorConfig, app);
    }
    
    useGen(isUsingGen) {
        this.client.setIsUsingGen(isUsingGen);
    }
    
    executeGraphql(query, options) {
        return this.client.executeGraphql(query, options);
    }
    
    executeGraphqlRead(query, options) {
        return this.client.executeGraphqlRead(query, options);
    }
    
    insert(tableName, variables) {
        return this.client.insert(tableName, variables);
    }
    
    insertMany(tableName, variables) {
        return this.client.insertMany(tableName, variables);
    }
    
    upsert(tableName, variables) {
        return this.client.upsert(tableName, variables);
    }
    
    upsertMany(tableName, variables) {
        return this.client.upsertMany(tableName, variables);
    }
    executeQuery(name, variables, options) {
        return this.client.executeQuery(name, variables, options);
    }
    executeMutation(name, variables, options) {
        return this.client.executeMutation(name, variables, options);
    }
}
exports.DataConnect = DataConnect;
