// firebase-admin/lib/data-connect/data-connect-api-client-internal.js

"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseDataConnectError = exports.DATA_CONNECT_ERROR_CODE_MAPPING = exports.DataConnectApiClient = void 0;
exports.useEmulator = useEmulator;
const api_request_1 = require("../utils/api-request");
const error_1 = require("../utils/error");
const utils = require("../utils/index");
const validator = require("../utils/validator");
const API_VERSION = 'v1';
const FIREBASE_DATA_CONNECT_PROD_URL = 'https://firebasedataconnect.googleapis.com';

const FIREBASE_DATA_CONNECT_SERVICES_URL_FORMAT = FIREBASE_DATA_CONNECT_PROD_URL +
    '/{version}' +
    '/projects/{projectId}' +
    '/locations/{locationId}' +
    '/services/{serviceId}' +
    ':{endpointId}';

const FIREBASE_DATA_CONNECT_CONNECTORS_URL_FORMAT = FIREBASE_DATA_CONNECT_PROD_URL +
    '/{version}' +
    '/projects/{projectId}' +
    '/locations/{locationId}' +
    '/services/{serviceId}' +
    '/connectors/{connectorId}' +
    ':{endpointId}';

const FIREBASE_DATA_CONNECT_EMULATOR_SERVICES_URL_FORMAT = 'http://{host}/{version}/projects/{projectId}/locations/{locationId}/services/{serviceId}:{endpointId}';

const FIREBASE_DATA_CONNECT_EMULATOR_CONNECTORS_URL_FORMAT = 'http://{host}/{version}/projects/{projectId}/locations/{locationId}/services/{serviceId}/connectors/{connectorId}:{endpointId}';
const EXECUTE_GRAPH_QL_ENDPOINT = 'executeGraphql';
const EXECUTE_GRAPH_QL_READ_ENDPOINT = 'executeGraphqlRead';
const IMPERSONATE_QUERY_ENDPOINT = 'impersonateQuery';
const IMPERSONATE_MUTATION_ENDPOINT = 'impersonateMutation';
function getHeaders(isUsingGen) {
    const headerValue = {
        'X-Firebase-Client': `fire-admin-node/${utils.getSdkVersion()}`,
        'X-Goog-Api-Client': utils.getMetricsHeader(),
    };
    if (isUsingGen) {
        headerValue['X-Goog-Api-Client'] += ' admin-js/gen';
    }
    return headerValue;
}

class DataConnectApiClient {
    constructor(connectorConfig, app) {
        this.connectorConfig = connectorConfig;
        this.app = app;
        this.isUsingGen = false;
        if (!validator.isNonNullObject(app) || !('options' in app)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, 'First argument passed to getDataConnect() must be a valid Firebase app instance.');
        }
        this.httpClient = new DataConnectHttpClient(app);
    }
    
    setIsUsingGen(isUsingGen) {
        this.isUsingGen = isUsingGen;
    }
    
    async executeGraphql(query, options) {
        return this.executeGraphqlHelper(query, EXECUTE_GRAPH_QL_ENDPOINT, options);
    }
    
    async executeGraphqlRead(query, options) {
        return this.executeGraphqlHelper(query, EXECUTE_GRAPH_QL_READ_ENDPOINT, options);
    }
    
    async executeGraphqlHelper(query, endpoint, options) {
        if (!validator.isNonEmptyString(query)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`query` must be a non-empty string.');
        }
        if (typeof options !== 'undefined') {
            if (!validator.isNonNullObject(options)) {
                throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, 'GraphqlOptions must be a non-null object');
            }
        }
        const data = {
            query,
            ...(options?.variables && { variables: options?.variables }),
            ...(options?.operationName && { operationName: options?.operationName }),
            ...(options?.impersonate && { extensions: { impersonate: options?.impersonate } }),
        };
        const url = await this.getServicesUrl(API_VERSION, this.connectorConfig.location, this.connectorConfig.serviceId, endpoint);
        try {
            const resp = await this.makeGqlRequest(url, data);
            return resp;
        }
        catch (err) {
            throw this.toFirebaseError(err);
        }
    }
    
    async executeQuery(name, variables, options) {
        return this.executeOperationHelper(IMPERSONATE_QUERY_ENDPOINT, name, variables, options);
    }
    
    async executeMutation(name, variables, options) {
        return this.executeOperationHelper(IMPERSONATE_MUTATION_ENDPOINT, name, variables, options);
    }
    
    async executeOperationHelper(endpoint, name, variables, options) {
        if (typeof name === 'undefined' ||
            !validator.isNonEmptyString(name)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`name` must be a non-empty string.');
        }
        if (this.connectorConfig.connector === undefined || this.connectorConfig.connector === '') {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, `The 'connectorConfig.connector' field used to instantiate your Data Connect
        instance must be a non-empty string (the connectorId) when calling executeQuery or executeMutation.`);
        }
        const data = {
            ...(variables && { variables: variables }),
            operationName: name,
            extensions: { impersonate: options?.impersonate },
        };
        const url = await this.getConnectorsUrl(API_VERSION, this.connectorConfig.location, this.connectorConfig.serviceId, this.connectorConfig.connector, endpoint);
        try {
            const resp = await this.makeGqlRequest(url, data);
            return resp;
        }
        catch (err) {
            throw this.toFirebaseError(err);
        }
    }
    
    async getServicesUrl(version, locationId, serviceId, endpointId) {
        const projectId = await this.getProjectId();
        const params = {
            version,
            projectId,
            locationId,
            serviceId,
            endpointId,
        };
        let urlFormat = FIREBASE_DATA_CONNECT_SERVICES_URL_FORMAT;
        if (useEmulator()) {
            urlFormat = FIREBASE_DATA_CONNECT_EMULATOR_SERVICES_URL_FORMAT;
            params.host = emulatorHost();
        }
        return utils.formatString(urlFormat, params);
    }
    
    async getConnectorsUrl(version, locationId, serviceId, connectorId, endpointId) {
        const projectId = await this.getProjectId();
        const params = {
            version,
            projectId,
            locationId,
            serviceId,
            connectorId,
            endpointId,
        };
        let urlFormat = FIREBASE_DATA_CONNECT_CONNECTORS_URL_FORMAT;
        if (useEmulator()) {
            urlFormat = FIREBASE_DATA_CONNECT_EMULATOR_CONNECTORS_URL_FORMAT;
            params.host = emulatorHost();
        }
        return utils.formatString(urlFormat, params);
    }
    getProjectId() {
        if (this.projectId) {
            return Promise.resolve(this.projectId);
        }
        return utils.findProjectId(this.app)
            .then((projectId) => {
            if (!validator.isNonEmptyString(projectId)) {
                throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.UNKNOWN, 'Failed to determine project ID. Initialize the '
                    + 'SDK with service account credentials or set project ID as an app option. '
                    + 'Alternatively, set the GOOGLE_CLOUD_PROJECT environment variable.');
            }
            this.projectId = projectId;
            return projectId;
        });
    }
    
    async makeGqlRequest(url, data) {
        const request = {
            method: 'POST',
            url,
            headers: getHeaders(this.isUsingGen),
            data,
        };
        const resp = await this.httpClient.send(request);
        if (resp.data.errors && validator.isNonEmptyArray(resp.data.errors)) {
            const allMessages = resp.data.errors.map((error) => error.message).join(' ');
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.QUERY_ERROR, allMessages);
        }
        return Promise.resolve({
            data: resp.data.data,
        });
    }
    toFirebaseError(err) {
        if (err instanceof error_1.PrefixedFirebaseError) {
            return err;
        }
        const response = err.response;
        if (!response.isJson()) {
            return new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.UNKNOWN, `Unexpected response with status: ${response.status} and body: ${response.text}`);
        }
        const error = response.data.error || {};
        let code = exports.DATA_CONNECT_ERROR_CODE_MAPPING.UNKNOWN;
        if (error.status && error.status in exports.DATA_CONNECT_ERROR_CODE_MAPPING) {
            code = exports.DATA_CONNECT_ERROR_CODE_MAPPING[error.status];
        }
        const message = error.message || `Unknown server error: ${response.text}`;
        return new FirebaseDataConnectError(code, message);
    }
    
    objectToString(data) {
        if (typeof data === 'string') {
            const escapedString = data
                .replace(/\\/g, '\\\\') // Replace \ with \\
                .replace(/"/g, '\\"'); // Replace " with \"
            return `"${escapedString}"`;
        }
        if (typeof data === 'number' || typeof data === 'boolean' || data === null) {
            return String(data);
        }
        if (validator.isArray(data)) {
            const elements = data.map(item => this.objectToString(item)).join(', ');
            return `[${elements}]`;
        }
        if (typeof data === 'object' && data !== null) {

            const kvPairs = Object.entries(data)
                .filter(([, val]) => val !== undefined)
                .map(([key, val]) => {

                return `${key}: ${this.objectToString(val)}`;
            });
            if (kvPairs.length === 0) {
                return '{}'; // Represent an object with no defined properties as {}
            }
            return `{ ${kvPairs.join(', ')} }`;
        }



        if (typeof data === 'undefined') {
            return 'null';
        }



        return String(data);
    }
    formatTableName(tableName) {

        if (tableName && tableName.length > 0) {
            return tableName.charAt(0).toLowerCase() + tableName.slice(1);
        }
        return tableName;
    }
    handleBulkImportErrors(err) {
        if (err.code === `data-connect/${exports.DATA_CONNECT_ERROR_CODE_MAPPING.QUERY_ERROR}`) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.QUERY_ERROR, `${err.message}. Make sure that your table name passed in matches the type name in your GraphQL schema file.`);
        }
        throw err;
    }
    
    async insert(tableName, data) {
        if (!validator.isNonEmptyString(tableName)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`tableName` must be a non-empty string.');
        }
        if (validator.isArray(data)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`data` must be an object, not an array, for single insert. For arrays, please use `insertMany` function.');
        }
        if (!validator.isNonNullObject(data)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`data` must be a non-null object.');
        }
        try {
            tableName = this.formatTableName(tableName);
            const gqlDataString = this.objectToString(data);
            const mutation = `mutation { ${tableName}_insert(data: ${gqlDataString}) }`;

            return this.executeGraphql(mutation).catch(this.handleBulkImportErrors);
        }
        catch (e) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INTERNAL, `Failed to construct insert mutation: ${e.message}`);
        }
    }
    
    async insertMany(tableName, data) {
        if (!validator.isNonEmptyString(tableName)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`tableName` must be a non-empty string.');
        }
        if (!validator.isNonEmptyArray(data)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`data` must be a non-empty array for insertMany.');
        }
        try {
            tableName = this.formatTableName(tableName);
            const gqlDataString = this.objectToString(data);
            const mutation = `mutation { ${tableName}_insertMany(data: ${gqlDataString}) }`;

            return this.executeGraphql(mutation).catch(this.handleBulkImportErrors);
        }
        catch (e) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INTERNAL, `Failed to construct insertMany mutation: ${e.message}`);
        }
    }
    
    async upsert(tableName, data) {
        if (!validator.isNonEmptyString(tableName)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`tableName` must be a non-empty string.');
        }
        if (validator.isArray(data)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`data` must be an object, not an array, for single upsert. For arrays, please use `upsertMany` function.');
        }
        if (!validator.isNonNullObject(data)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`data` must be a non-null object.');
        }
        try {
            tableName = this.formatTableName(tableName);
            const gqlDataString = this.objectToString(data);
            const mutation = `mutation { ${tableName}_upsert(data: ${gqlDataString}) }`;

            return this.executeGraphql(mutation).catch(this.handleBulkImportErrors);
        }
        catch (e) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INTERNAL, `Failed to construct upsert mutation: ${e.message}`);
        }
    }
    
    async upsertMany(tableName, data) {
        if (!validator.isNonEmptyString(tableName)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`tableName` must be a non-empty string.');
        }
        if (!validator.isNonEmptyArray(data)) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT, '`data` must be a non-empty array for upsertMany.');
        }
        try {
            tableName = this.formatTableName(tableName);
            const gqlDataString = this.objectToString(data);
            const mutation = `mutation { ${tableName}_upsertMany(data: ${gqlDataString}) }`;

            return this.executeGraphql(mutation).catch(this.handleBulkImportErrors);
        }
        catch (e) {
            throw new FirebaseDataConnectError(exports.DATA_CONNECT_ERROR_CODE_MAPPING.INTERNAL, `Failed to construct upsertMany mutation: ${e.message}`);
        }
    }
}
exports.DataConnectApiClient = DataConnectApiClient;

class DataConnectHttpClient extends api_request_1.AuthorizedHttpClient {
    getToken() {
        if (useEmulator()) {
            return Promise.resolve('owner');
        }
        return super.getToken();
    }
}
function emulatorHost() {
    return process.env.DATA_CONNECT_EMULATOR_HOST;
}

function useEmulator() {
    return !!emulatorHost();
}
exports.DATA_CONNECT_ERROR_CODE_MAPPING = {
    ABORTED: 'aborted',
    INVALID_ARGUMENT: 'invalid-argument',
    INVALID_CREDENTIAL: 'invalid-credential',
    INTERNAL: 'internal-error',
    PERMISSION_DENIED: 'permission-denied',
    UNAUTHENTICATED: 'unauthenticated',
    NOT_FOUND: 'not-found',
    UNKNOWN: 'unknown-error',
    QUERY_ERROR: 'query-error',
};

class FirebaseDataConnectError extends error_1.PrefixedFirebaseError {
    constructor(code, message) {
        super('data-connect', code, message);
        


        
        this.__proto__ = FirebaseDataConnectError.prototype;
    }
}
exports.FirebaseDataConnectError = FirebaseDataConnectError;
