// firebase-functions/lib/common/change.js



function applyFieldMask(sparseBefore, after, fieldMask) {
	const before = { ...after };
	const masks = fieldMask.split(",");
	for (const mask of masks) {
		const parts = mask.split(".");
		const head = parts[0];
		const tail = parts.slice(1).join(".");
		if (parts.length > 1) {
			before[head] = applyFieldMask(sparseBefore?.[head], after[head], tail);
			continue;
		}
		const val = sparseBefore?.[head];
		if (typeof val === "undefined") {
			delete before[mask];
		} else {
			before[mask] = val;
		}
	}
	return before;
}

var Change = class Change {
	
	static fromObjects(before, after) {
		return new Change(before, after);
	}
	
	static fromJSON(json, customizer = (x) => x) {
		let before = { ...json.before };
		if (json.fieldMask) {
			before = applyFieldMask(before, json.after, json.fieldMask);
		}
		return Change.fromObjects(customizer(before || {}), customizer(json.after || {}));
	}
	constructor(before, after) {
		this.before = before;
		this.after = after;
	}
};


exports.Change = Change;
exports.applyFieldMask = applyFieldMask;
