// firebase-functions/lib/common/trace.js
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.js');
let async_hooks = require("async_hooks");
async_hooks = require_rolldown_runtime.__toESM(async_hooks);


const traceContext = new async_hooks.AsyncLocalStorage();

const CLOUD_TRACE_REGEX = new RegExp("^(?<traceId>[A-Fa-f0-9]{32})/" + "(?<parentIdInt>[0-9]+)" + "(?:;o=(?<traceMask>[0-3]))?$");
const CLOUD_TRACE_HEADER = "X-Cloud-Trace-Context";
function matchCloudTraceHeader(carrier) {
	let header = carrier?.[CLOUD_TRACE_HEADER];
	if (!header) {
		header = carrier?.[CLOUD_TRACE_HEADER.toLowerCase()];
	}
	if (header && typeof header === "string") {
		const matches = CLOUD_TRACE_REGEX.exec(header);
		if (matches && matches.groups) {
			const { traceId, parentIdInt, traceMask } = matches.groups;
			const parentId = parseInt(parentIdInt);
			if (isNaN(parentId)) {
				return;
			}
			const sample = !!traceMask && traceMask !== "0";
			return {
				traceId,
				parentId: parentId.toString(16),
				sample,
				version: "00"
			};
		}
	}
}

const TRACEPARENT_REGEX = new RegExp("^(?<version>[a-f0-9]{2})-" + "(?<traceId>[a-f0-9]{32})-" + "(?<parentId>[a-f0-9]{16})-" + "(?<flag>[a-f0-9]{2})$");
const TRACEPARENT_HEADER = "traceparent";
function matchTraceparentHeader(carrier) {
	const header = carrier?.[TRACEPARENT_HEADER];
	if (header && typeof header === "string") {
		const matches = TRACEPARENT_REGEX.exec(header);
		if (matches && matches.groups) {
			const { version, traceId, parentId, flag } = matches.groups;
			const sample = flag === "01";
			return {
				traceId,
				parentId,
				sample,
				version
			};
		}
	}
}

function extractTraceContext(carrier) {
	return matchCloudTraceHeader(carrier) || matchTraceparentHeader(carrier);
}


exports.extractTraceContext = extractTraceContext;
exports.traceContext = traceContext;
