// @opentelemetry/api/build/esm/trace/NonRecordingSpan.js

import { INVALID_SPAN_CONTEXT } from './invalid-span-constants';

var NonRecordingSpan =  (function () {
    function NonRecordingSpan(_spanContext) {
        if (_spanContext === void 0) { _spanContext = INVALID_SPAN_CONTEXT; }
        this._spanContext = _spanContext;
    }

    NonRecordingSpan.prototype.spanContext = function () {
        return this._spanContext;
    };

    NonRecordingSpan.prototype.setAttribute = function (_key, _value) {
        return this;
    };

    NonRecordingSpan.prototype.setAttributes = function (_attributes) {
        return this;
    };

    NonRecordingSpan.prototype.addEvent = function (_name, _attributes) {
        return this;
    };
    NonRecordingSpan.prototype.addLink = function (_link) {
        return this;
    };
    NonRecordingSpan.prototype.addLinks = function (_links) {
        return this;
    };

    NonRecordingSpan.prototype.setStatus = function (_status) {
        return this;
    };

    NonRecordingSpan.prototype.updateName = function (_name) {
        return this;
    };

    NonRecordingSpan.prototype.end = function (_endTime) { };

    NonRecordingSpan.prototype.isRecording = function () {
        return false;
    };

    NonRecordingSpan.prototype.recordException = function (_exception, _time) { };
    return NonRecordingSpan;
}());
export { NonRecordingSpan };
