// @opentelemetry/api/build/esnext/metrics/NoopMeter.js


export class NoopMeter {
    constructor() { }
    
    createGauge(_name, _options) {
        return NOOP_GAUGE_METRIC;
    }
    
    createHistogram(_name, _options) {
        return NOOP_HISTOGRAM_METRIC;
    }
    
    createCounter(_name, _options) {
        return NOOP_COUNTER_METRIC;
    }
    
    createUpDownCounter(_name, _options) {
        return NOOP_UP_DOWN_COUNTER_METRIC;
    }
    
    createObservableGauge(_name, _options) {
        return NOOP_OBSERVABLE_GAUGE_METRIC;
    }
    
    createObservableCounter(_name, _options) {
        return NOOP_OBSERVABLE_COUNTER_METRIC;
    }
    
    createObservableUpDownCounter(_name, _options) {
        return NOOP_OBSERVABLE_UP_DOWN_COUNTER_METRIC;
    }
    
    addBatchObservableCallback(_callback, _observables) { }
    
    removeBatchObservableCallback(_callback) { }
}
export class NoopMetric {
}
export class NoopCounterMetric extends NoopMetric {
    add(_value, _attributes) { }
}
export class NoopUpDownCounterMetric extends NoopMetric {
    add(_value, _attributes) { }
}
export class NoopGaugeMetric extends NoopMetric {
    record(_value, _attributes) { }
}
export class NoopHistogramMetric extends NoopMetric {
    record(_value, _attributes) { }
}
export class NoopObservableMetric {
    addCallback(_callback) { }
    removeCallback(_callback) { }
}
export class NoopObservableCounterMetric extends NoopObservableMetric {
}
export class NoopObservableGaugeMetric extends NoopObservableMetric {
}
export class NoopObservableUpDownCounterMetric extends NoopObservableMetric {
}
export const NOOP_METER = new NoopMeter();

export const NOOP_COUNTER_METRIC = new NoopCounterMetric();
export const NOOP_GAUGE_METRIC = new NoopGaugeMetric();
export const NOOP_HISTOGRAM_METRIC = new NoopHistogramMetric();
export const NOOP_UP_DOWN_COUNTER_METRIC = new NoopUpDownCounterMetric();

export const NOOP_OBSERVABLE_COUNTER_METRIC = new NoopObservableCounterMetric();
export const NOOP_OBSERVABLE_GAUGE_METRIC = new NoopObservableGaugeMetric();
export const NOOP_OBSERVABLE_UP_DOWN_COUNTER_METRIC = new NoopObservableUpDownCounterMetric();

export function createNoopMeter() {
    return NOOP_METER;
}
