// firebase-functions/protos/compiledFirestore.js

"use strict";

var $protobuf = require("protobufjs/minimal");


var $Reader = $protobuf.Reader, $Writer = $protobuf.Writer, $util = $protobuf.util;


var $root = $protobuf.roots["default"] || ($protobuf.roots["default"] = {});

$root.google = (function() {

    
    var google = {};

    google.protobuf = (function() {

        
        var protobuf = {};

        protobuf.Struct = (function() {

            

            
            function Struct(properties) {
                this.fields = {};
                if (properties)
                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                        if (properties[keys[i]] != null)
                            this[keys[i]] = properties[keys[i]];
            }

            
            Struct.prototype.fields = $util.emptyObject;

            
            Struct.create = function create(properties) {
                return new Struct(properties);
            };

            
            Struct.encode = function encode(message, writer) {
                if (!writer)
                    writer = $Writer.create();
                if (message.fields != null && Object.hasOwnProperty.call(message, "fields"))
                    for (var keys = Object.keys(message.fields), i = 0; i < keys.length; ++i) {
                        writer.uint32(10).fork().uint32(10).string(keys[i]);
                        $root.google.protobuf.Value.encode(message.fields[keys[i]], writer.uint32(18).fork()).ldelim().ldelim();
                    }
                return writer;
            };

            
            Struct.encodeDelimited = function encodeDelimited(message, writer) {
                return this.encode(message, writer).ldelim();
            };

            
            Struct.decode = function decode(reader, length) {
                if (!(reader instanceof $Reader))
                    reader = $Reader.create(reader);
                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.protobuf.Struct(), key, value;
                while (reader.pos < end) {
                    var tag = reader.uint32();
                    switch (tag >>> 3) {
                    case 1: {
                            if (message.fields === $util.emptyObject)
                                message.fields = {};
                            var end2 = reader.uint32() + reader.pos;
                            key = "";
                            value = null;
                            while (reader.pos < end2) {
                                var tag2 = reader.uint32();
                                switch (tag2 >>> 3) {
                                case 1:
                                    key = reader.string();
                                    break;
                                case 2:
                                    value = $root.google.protobuf.Value.decode(reader, reader.uint32());
                                    break;
                                default:
                                    reader.skipType(tag2 & 7);
                                    break;
                                }
                            }
                            message.fields[key] = value;
                            break;
                        }
                    default:
                        reader.skipType(tag & 7);
                        break;
                    }
                }
                return message;
            };

            
            Struct.decodeDelimited = function decodeDelimited(reader) {
                if (!(reader instanceof $Reader))
                    reader = new $Reader(reader);
                return this.decode(reader, reader.uint32());
            };

            
            Struct.verify = function verify(message) {
                if (typeof message !== "object" || message === null)
                    return "object expected";
                if (message.fields != null && message.hasOwnProperty("fields")) {
                    if (!$util.isObject(message.fields))
                        return "fields: object expected";
                    var key = Object.keys(message.fields);
                    for (var i = 0; i < key.length; ++i) {
                        var error = $root.google.protobuf.Value.verify(message.fields[key[i]]);
                        if (error)
                            return "fields." + error;
                    }
                }
                return null;
            };

            
            Struct.fromObject = function fromObject(object) {
                if (object instanceof $root.google.protobuf.Struct)
                    return object;
                var message = new $root.google.protobuf.Struct();
                if (object.fields) {
                    if (typeof object.fields !== "object")
                        throw TypeError(".google.protobuf.Struct.fields: object expected");
                    message.fields = {};
                    for (var keys = Object.keys(object.fields), i = 0; i < keys.length; ++i) {
                        if (typeof object.fields[keys[i]] !== "object")
                            throw TypeError(".google.protobuf.Struct.fields: object expected");
                        message.fields[keys[i]] = $root.google.protobuf.Value.fromObject(object.fields[keys[i]]);
                    }
                }
                return message;
            };

            
            Struct.toObject = function toObject(message, options) {
                if (!options)
                    options = {};
                var object = {};
                if (options.objects || options.defaults)
                    object.fields = {};
                var keys2;
                if (message.fields && (keys2 = Object.keys(message.fields)).length) {
                    object.fields = {};
                    for (var j = 0; j < keys2.length; ++j)
                        object.fields[keys2[j]] = $root.google.protobuf.Value.toObject(message.fields[keys2[j]], options);
                }
                return object;
            };

            
            Struct.prototype.toJSON = function toJSON() {
                return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
            };

            
            Struct.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                if (typeUrlPrefix === undefined) {
                    typeUrlPrefix = "type.googleapis.com";
                }
                return typeUrlPrefix + "/google.protobuf.Struct";
            };

            return Struct;
        })();

        protobuf.Value = (function() {

            

            
            function Value(properties) {
                if (properties)
                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                        if (properties[keys[i]] != null)
                            this[keys[i]] = properties[keys[i]];
            }

            
            Value.prototype.nullValue = null;

            
            Value.prototype.numberValue = null;

            
            Value.prototype.stringValue = null;

            
            Value.prototype.boolValue = null;

            
            Value.prototype.structValue = null;

            
            Value.prototype.listValue = null;


            var $oneOfFields;

            
            Object.defineProperty(Value.prototype, "kind", {
                get: $util.oneOfGetter($oneOfFields = ["nullValue", "numberValue", "stringValue", "boolValue", "structValue", "listValue"]),
                set: $util.oneOfSetter($oneOfFields)
            });

            
            Value.create = function create(properties) {
                return new Value(properties);
            };

            
            Value.encode = function encode(message, writer) {
                if (!writer)
                    writer = $Writer.create();
                if (message.nullValue != null && Object.hasOwnProperty.call(message, "nullValue"))
                    writer.uint32(8).int32(message.nullValue);
                if (message.numberValue != null && Object.hasOwnProperty.call(message, "numberValue"))
                    writer.uint32(17).double(message.numberValue);
                if (message.stringValue != null && Object.hasOwnProperty.call(message, "stringValue"))
                    writer.uint32(26).string(message.stringValue);
                if (message.boolValue != null && Object.hasOwnProperty.call(message, "boolValue"))
                    writer.uint32(32).bool(message.boolValue);
                if (message.structValue != null && Object.hasOwnProperty.call(message, "structValue"))
                    $root.google.protobuf.Struct.encode(message.structValue, writer.uint32(42).fork()).ldelim();
                if (message.listValue != null && Object.hasOwnProperty.call(message, "listValue"))
                    $root.google.protobuf.ListValue.encode(message.listValue, writer.uint32(50).fork()).ldelim();
                return writer;
            };

            
            Value.encodeDelimited = function encodeDelimited(message, writer) {
                return this.encode(message, writer).ldelim();
            };

            
            Value.decode = function decode(reader, length) {
                if (!(reader instanceof $Reader))
                    reader = $Reader.create(reader);
                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.protobuf.Value();
                while (reader.pos < end) {
                    var tag = reader.uint32();
                    switch (tag >>> 3) {
                    case 1: {
                            message.nullValue = reader.int32();
                            break;
                        }
                    case 2: {
                            message.numberValue = reader.double();
                            break;
                        }
                    case 3: {
                            message.stringValue = reader.string();
                            break;
                        }
                    case 4: {
                            message.boolValue = reader.bool();
                            break;
                        }
                    case 5: {
                            message.structValue = $root.google.protobuf.Struct.decode(reader, reader.uint32());
                            break;
                        }
                    case 6: {
                            message.listValue = $root.google.protobuf.ListValue.decode(reader, reader.uint32());
                            break;
                        }
                    default:
                        reader.skipType(tag & 7);
                        break;
                    }
                }
                return message;
            };

            
            Value.decodeDelimited = function decodeDelimited(reader) {
                if (!(reader instanceof $Reader))
                    reader = new $Reader(reader);
                return this.decode(reader, reader.uint32());
            };

            
            Value.verify = function verify(message) {
                if (typeof message !== "object" || message === null)
                    return "object expected";
                var properties = {};
                if (message.nullValue != null && message.hasOwnProperty("nullValue")) {
                    properties.kind = 1;
                    switch (message.nullValue) {
                    default:
                        return "nullValue: enum value expected";
                    case 0:
                        break;
                    }
                }
                if (message.numberValue != null && message.hasOwnProperty("numberValue")) {
                    if (properties.kind === 1)
                        return "kind: multiple values";
                    properties.kind = 1;
                    if (typeof message.numberValue !== "number")
                        return "numberValue: number expected";
                }
                if (message.stringValue != null && message.hasOwnProperty("stringValue")) {
                    if (properties.kind === 1)
                        return "kind: multiple values";
                    properties.kind = 1;
                    if (!$util.isString(message.stringValue))
                        return "stringValue: string expected";
                }
                if (message.boolValue != null && message.hasOwnProperty("boolValue")) {
                    if (properties.kind === 1)
                        return "kind: multiple values";
                    properties.kind = 1;
                    if (typeof message.boolValue !== "boolean")
                        return "boolValue: boolean expected";
                }
                if (message.structValue != null && message.hasOwnProperty("structValue")) {
                    if (properties.kind === 1)
                        return "kind: multiple values";
                    properties.kind = 1;
                    {
                        var error = $root.google.protobuf.Struct.verify(message.structValue);
                        if (error)
                            return "structValue." + error;
                    }
                }
                if (message.listValue != null && message.hasOwnProperty("listValue")) {
                    if (properties.kind === 1)
                        return "kind: multiple values";
                    properties.kind = 1;
                    {
                        var error = $root.google.protobuf.ListValue.verify(message.listValue);
                        if (error)
                            return "listValue." + error;
                    }
                }
                return null;
            };

            
            Value.fromObject = function fromObject(object) {
                if (object instanceof $root.google.protobuf.Value)
                    return object;
                var message = new $root.google.protobuf.Value();
                switch (object.nullValue) {
                default:
                    if (typeof object.nullValue === "number") {
                        message.nullValue = object.nullValue;
                        break;
                    }
                    break;
                case "NULL_VALUE":
                case 0:
                    message.nullValue = 0;
                    break;
                }
                if (object.numberValue != null)
                    message.numberValue = Number(object.numberValue);
                if (object.stringValue != null)
                    message.stringValue = String(object.stringValue);
                if (object.boolValue != null)
                    message.boolValue = Boolean(object.boolValue);
                if (object.structValue != null) {
                    if (typeof object.structValue !== "object")
                        throw TypeError(".google.protobuf.Value.structValue: object expected");
                    message.structValue = $root.google.protobuf.Struct.fromObject(object.structValue);
                }
                if (object.listValue != null) {
                    if (typeof object.listValue !== "object")
                        throw TypeError(".google.protobuf.Value.listValue: object expected");
                    message.listValue = $root.google.protobuf.ListValue.fromObject(object.listValue);
                }
                return message;
            };

            
            Value.toObject = function toObject(message, options) {
                if (!options)
                    options = {};
                var object = {};
                if (message.nullValue != null && message.hasOwnProperty("nullValue")) {
                    object.nullValue = options.enums === String ? $root.google.protobuf.NullValue[message.nullValue] === undefined ? message.nullValue : $root.google.protobuf.NullValue[message.nullValue] : message.nullValue;
                    if (options.oneofs)
                        object.kind = "nullValue";
                }
                if (message.numberValue != null && message.hasOwnProperty("numberValue")) {
                    object.numberValue = options.json && !isFinite(message.numberValue) ? String(message.numberValue) : message.numberValue;
                    if (options.oneofs)
                        object.kind = "numberValue";
                }
                if (message.stringValue != null && message.hasOwnProperty("stringValue")) {
                    object.stringValue = message.stringValue;
                    if (options.oneofs)
                        object.kind = "stringValue";
                }
                if (message.boolValue != null && message.hasOwnProperty("boolValue")) {
                    object.boolValue = message.boolValue;
                    if (options.oneofs)
                        object.kind = "boolValue";
                }
                if (message.structValue != null && message.hasOwnProperty("structValue")) {
                    object.structValue = $root.google.protobuf.Struct.toObject(message.structValue, options);
                    if (options.oneofs)
                        object.kind = "structValue";
                }
                if (message.listValue != null && message.hasOwnProperty("listValue")) {
                    object.listValue = $root.google.protobuf.ListValue.toObject(message.listValue, options);
                    if (options.oneofs)
                        object.kind = "listValue";
                }
                return object;
            };

            
            Value.prototype.toJSON = function toJSON() {
                return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
            };

            
            Value.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                if (typeUrlPrefix === undefined) {
                    typeUrlPrefix = "type.googleapis.com";
                }
                return typeUrlPrefix + "/google.protobuf.Value";
            };

            return Value;
        })();

        
        protobuf.NullValue = (function() {
            var valuesById = {}, values = Object.create(valuesById);
            values[valuesById[0] = "NULL_VALUE"] = 0;
            return values;
        })();

        protobuf.ListValue = (function() {

            

            
            function ListValue(properties) {
                this.values = [];
                if (properties)
                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                        if (properties[keys[i]] != null)
                            this[keys[i]] = properties[keys[i]];
            }

            
            ListValue.prototype.values = $util.emptyArray;

            
            ListValue.create = function create(properties) {
                return new ListValue(properties);
            };

            
            ListValue.encode = function encode(message, writer) {
                if (!writer)
                    writer = $Writer.create();
                if (message.values != null && message.values.length)
                    for (var i = 0; i < message.values.length; ++i)
                        $root.google.protobuf.Value.encode(message.values[i], writer.uint32(10).fork()).ldelim();
                return writer;
            };

            
            ListValue.encodeDelimited = function encodeDelimited(message, writer) {
                return this.encode(message, writer).ldelim();
            };

            
            ListValue.decode = function decode(reader, length) {
                if (!(reader instanceof $Reader))
                    reader = $Reader.create(reader);
                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.protobuf.ListValue();
                while (reader.pos < end) {
                    var tag = reader.uint32();
                    switch (tag >>> 3) {
                    case 1: {
                            if (!(message.values && message.values.length))
                                message.values = [];
                            message.values.push($root.google.protobuf.Value.decode(reader, reader.uint32()));
                            break;
                        }
                    default:
                        reader.skipType(tag & 7);
                        break;
                    }
                }
                return message;
            };

            
            ListValue.decodeDelimited = function decodeDelimited(reader) {
                if (!(reader instanceof $Reader))
                    reader = new $Reader(reader);
                return this.decode(reader, reader.uint32());
            };

            
            ListValue.verify = function verify(message) {
                if (typeof message !== "object" || message === null)
                    return "object expected";
                if (message.values != null && message.hasOwnProperty("values")) {
                    if (!Array.isArray(message.values))
                        return "values: array expected";
                    for (var i = 0; i < message.values.length; ++i) {
                        var error = $root.google.protobuf.Value.verify(message.values[i]);
                        if (error)
                            return "values." + error;
                    }
                }
                return null;
            };

            
            ListValue.fromObject = function fromObject(object) {
                if (object instanceof $root.google.protobuf.ListValue)
                    return object;
                var message = new $root.google.protobuf.ListValue();
                if (object.values) {
                    if (!Array.isArray(object.values))
                        throw TypeError(".google.protobuf.ListValue.values: array expected");
                    message.values = [];
                    for (var i = 0; i < object.values.length; ++i) {
                        if (typeof object.values[i] !== "object")
                            throw TypeError(".google.protobuf.ListValue.values: object expected");
                        message.values[i] = $root.google.protobuf.Value.fromObject(object.values[i]);
                    }
                }
                return message;
            };

            
            ListValue.toObject = function toObject(message, options) {
                if (!options)
                    options = {};
                var object = {};
                if (options.arrays || options.defaults)
                    object.values = [];
                if (message.values && message.values.length) {
                    object.values = [];
                    for (var j = 0; j < message.values.length; ++j)
                        object.values[j] = $root.google.protobuf.Value.toObject(message.values[j], options);
                }
                return object;
            };

            
            ListValue.prototype.toJSON = function toJSON() {
                return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
            };

            
            ListValue.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                if (typeUrlPrefix === undefined) {
                    typeUrlPrefix = "type.googleapis.com";
                }
                return typeUrlPrefix + "/google.protobuf.ListValue";
            };

            return ListValue;
        })();

        protobuf.Timestamp = (function() {

            

            
            function Timestamp(properties) {
                if (properties)
                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                        if (properties[keys[i]] != null)
                            this[keys[i]] = properties[keys[i]];
            }

            
            Timestamp.prototype.seconds = $util.Long ? $util.Long.fromBits(0,0,false) : 0;

            
            Timestamp.prototype.nanos = 0;

            
            Timestamp.create = function create(properties) {
                return new Timestamp(properties);
            };

            
            Timestamp.encode = function encode(message, writer) {
                if (!writer)
                    writer = $Writer.create();
                if (message.seconds != null && Object.hasOwnProperty.call(message, "seconds"))
                    writer.uint32(8).int64(message.seconds);
                if (message.nanos != null && Object.hasOwnProperty.call(message, "nanos"))
                    writer.uint32(16).int32(message.nanos);
                return writer;
            };

            
            Timestamp.encodeDelimited = function encodeDelimited(message, writer) {
                return this.encode(message, writer).ldelim();
            };

            
            Timestamp.decode = function decode(reader, length) {
                if (!(reader instanceof $Reader))
                    reader = $Reader.create(reader);
                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.protobuf.Timestamp();
                while (reader.pos < end) {
                    var tag = reader.uint32();
                    switch (tag >>> 3) {
                    case 1: {
                            message.seconds = reader.int64();
                            break;
                        }
                    case 2: {
                            message.nanos = reader.int32();
                            break;
                        }
                    default:
                        reader.skipType(tag & 7);
                        break;
                    }
                }
                return message;
            };

            
            Timestamp.decodeDelimited = function decodeDelimited(reader) {
                if (!(reader instanceof $Reader))
                    reader = new $Reader(reader);
                return this.decode(reader, reader.uint32());
            };

            
            Timestamp.verify = function verify(message) {
                if (typeof message !== "object" || message === null)
                    return "object expected";
                if (message.seconds != null && message.hasOwnProperty("seconds"))
                    if (!$util.isInteger(message.seconds) && !(message.seconds && $util.isInteger(message.seconds.low) && $util.isInteger(message.seconds.high)))
                        return "seconds: integer|Long expected";
                if (message.nanos != null && message.hasOwnProperty("nanos"))
                    if (!$util.isInteger(message.nanos))
                        return "nanos: integer expected";
                return null;
            };

            
            Timestamp.fromObject = function fromObject(object) {
                if (object instanceof $root.google.protobuf.Timestamp)
                    return object;
                var message = new $root.google.protobuf.Timestamp();
                if (object.seconds != null)
                    if ($util.Long)
                        (message.seconds = $util.Long.fromValue(object.seconds)).unsigned = false;
                    else if (typeof object.seconds === "string")
                        message.seconds = parseInt(object.seconds, 10);
                    else if (typeof object.seconds === "number")
                        message.seconds = object.seconds;
                    else if (typeof object.seconds === "object")
                        message.seconds = new $util.LongBits(object.seconds.low >>> 0, object.seconds.high >>> 0).toNumber();
                if (object.nanos != null)
                    message.nanos = object.nanos | 0;
                return message;
            };

            
            Timestamp.toObject = function toObject(message, options) {
                if (!options)
                    options = {};
                var object = {};
                if (options.defaults) {
                    if ($util.Long) {
                        var long = new $util.Long(0, 0, false);
                        object.seconds = options.longs === String ? long.toString() : options.longs === Number ? long.toNumber() : long;
                    } else
                        object.seconds = options.longs === String ? "0" : 0;
                    object.nanos = 0;
                }
                if (message.seconds != null && message.hasOwnProperty("seconds"))
                    if (typeof message.seconds === "number")
                        object.seconds = options.longs === String ? String(message.seconds) : message.seconds;
                    else
                        object.seconds = options.longs === String ? $util.Long.prototype.toString.call(message.seconds) : options.longs === Number ? new $util.LongBits(message.seconds.low >>> 0, message.seconds.high >>> 0).toNumber() : message.seconds;
                if (message.nanos != null && message.hasOwnProperty("nanos"))
                    object.nanos = message.nanos;
                return object;
            };

            
            Timestamp.prototype.toJSON = function toJSON() {
                return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
            };

            
            Timestamp.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                if (typeUrlPrefix === undefined) {
                    typeUrlPrefix = "type.googleapis.com";
                }
                return typeUrlPrefix + "/google.protobuf.Timestamp";
            };

            return Timestamp;
        })();

        protobuf.Any = (function() {

            

            
            function Any(properties) {
                if (properties)
                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                        if (properties[keys[i]] != null)
                            this[keys[i]] = properties[keys[i]];
            }

            
            Any.prototype.typeUrl = "";

            
            Any.prototype.value = $util.newBuffer([]);

            
            Any.create = function create(properties) {
                return new Any(properties);
            };

            
            Any.encode = function encode(message, writer) {
                if (!writer)
                    writer = $Writer.create();
                if (message.typeUrl != null && Object.hasOwnProperty.call(message, "typeUrl"))
                    writer.uint32(10).string(message.typeUrl);
                if (message.value != null && Object.hasOwnProperty.call(message, "value"))
                    writer.uint32(18).bytes(message.value);
                return writer;
            };

            
            Any.encodeDelimited = function encodeDelimited(message, writer) {
                return this.encode(message, writer).ldelim();
            };

            
            Any.decode = function decode(reader, length) {
                if (!(reader instanceof $Reader))
                    reader = $Reader.create(reader);
                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.protobuf.Any();
                while (reader.pos < end) {
                    var tag = reader.uint32();
                    switch (tag >>> 3) {
                    case 1: {
                            message.typeUrl = reader.string();
                            break;
                        }
                    case 2: {
                            message.value = reader.bytes();
                            break;
                        }
                    default:
                        reader.skipType(tag & 7);
                        break;
                    }
                }
                return message;
            };

            
            Any.decodeDelimited = function decodeDelimited(reader) {
                if (!(reader instanceof $Reader))
                    reader = new $Reader(reader);
                return this.decode(reader, reader.uint32());
            };

            
            Any.verify = function verify(message) {
                if (typeof message !== "object" || message === null)
                    return "object expected";
                if (message.typeUrl != null && message.hasOwnProperty("typeUrl"))
                    if (!$util.isString(message.typeUrl))
                        return "typeUrl: string expected";
                if (message.value != null && message.hasOwnProperty("value"))
                    if (!(message.value && typeof message.value.length === "number" || $util.isString(message.value)))
                        return "value: buffer expected";
                return null;
            };

            
            Any.fromObject = function fromObject(object) {
                if (object instanceof $root.google.protobuf.Any)
                    return object;
                var message = new $root.google.protobuf.Any();
                if (object.typeUrl != null)
                    message.typeUrl = String(object.typeUrl);
                if (object.value != null)
                    if (typeof object.value === "string")
                        $util.base64.decode(object.value, message.value = $util.newBuffer($util.base64.length(object.value)), 0);
                    else if (object.value.length >= 0)
                        message.value = object.value;
                return message;
            };

            
            Any.toObject = function toObject(message, options) {
                if (!options)
                    options = {};
                var object = {};
                if (options.defaults) {
                    object.typeUrl = "";
                    if (options.bytes === String)
                        object.value = "";
                    else {
                        object.value = [];
                        if (options.bytes !== Array)
                            object.value = $util.newBuffer(object.value);
                    }
                }
                if (message.typeUrl != null && message.hasOwnProperty("typeUrl"))
                    object.typeUrl = message.typeUrl;
                if (message.value != null && message.hasOwnProperty("value"))
                    object.value = options.bytes === String ? $util.base64.encode(message.value, 0, message.value.length) : options.bytes === Array ? Array.prototype.slice.call(message.value) : message.value;
                return object;
            };

            
            Any.prototype.toJSON = function toJSON() {
                return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
            };

            
            Any.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                if (typeUrlPrefix === undefined) {
                    typeUrlPrefix = "type.googleapis.com";
                }
                return typeUrlPrefix + "/google.protobuf.Any";
            };

            return Any;
        })();

        return protobuf;
    })();

    google.events = (function() {

        
        var events = {};

        events.cloud = (function() {

            
            var cloud = {};

            cloud.firestore = (function() {

                
                var firestore = {};

                firestore.v1 = (function() {

                    
                    var v1 = {};

                    v1.DocumentEventData = (function() {

                        

                        
                        function DocumentEventData(properties) {
                            if (properties)
                                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                                    if (properties[keys[i]] != null)
                                        this[keys[i]] = properties[keys[i]];
                        }

                        
                        DocumentEventData.prototype.value = null;

                        
                        DocumentEventData.prototype.oldValue = null;

                        
                        DocumentEventData.prototype.updateMask = null;

                        
                        DocumentEventData.create = function create(properties) {
                            return new DocumentEventData(properties);
                        };

                        
                        DocumentEventData.encode = function encode(message, writer) {
                            if (!writer)
                                writer = $Writer.create();
                            if (message.value != null && Object.hasOwnProperty.call(message, "value"))
                                $root.google.events.cloud.firestore.v1.Document.encode(message.value, writer.uint32(10).fork()).ldelim();
                            if (message.oldValue != null && Object.hasOwnProperty.call(message, "oldValue"))
                                $root.google.events.cloud.firestore.v1.Document.encode(message.oldValue, writer.uint32(18).fork()).ldelim();
                            if (message.updateMask != null && Object.hasOwnProperty.call(message, "updateMask"))
                                $root.google.events.cloud.firestore.v1.DocumentMask.encode(message.updateMask, writer.uint32(26).fork()).ldelim();
                            return writer;
                        };

                        
                        DocumentEventData.encodeDelimited = function encodeDelimited(message, writer) {
                            return this.encode(message, writer).ldelim();
                        };

                        
                        DocumentEventData.decode = function decode(reader, length) {
                            if (!(reader instanceof $Reader))
                                reader = $Reader.create(reader);
                            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.events.cloud.firestore.v1.DocumentEventData();
                            while (reader.pos < end) {
                                var tag = reader.uint32();
                                switch (tag >>> 3) {
                                case 1: {
                                        message.value = $root.google.events.cloud.firestore.v1.Document.decode(reader, reader.uint32());
                                        break;
                                    }
                                case 2: {
                                        message.oldValue = $root.google.events.cloud.firestore.v1.Document.decode(reader, reader.uint32());
                                        break;
                                    }
                                case 3: {
                                        message.updateMask = $root.google.events.cloud.firestore.v1.DocumentMask.decode(reader, reader.uint32());
                                        break;
                                    }
                                default:
                                    reader.skipType(tag & 7);
                                    break;
                                }
                            }
                            return message;
                        };

                        
                        DocumentEventData.decodeDelimited = function decodeDelimited(reader) {
                            if (!(reader instanceof $Reader))
                                reader = new $Reader(reader);
                            return this.decode(reader, reader.uint32());
                        };

                        
                        DocumentEventData.verify = function verify(message) {
                            if (typeof message !== "object" || message === null)
                                return "object expected";
                            if (message.value != null && message.hasOwnProperty("value")) {
                                var error = $root.google.events.cloud.firestore.v1.Document.verify(message.value);
                                if (error)
                                    return "value." + error;
                            }
                            if (message.oldValue != null && message.hasOwnProperty("oldValue")) {
                                var error = $root.google.events.cloud.firestore.v1.Document.verify(message.oldValue);
                                if (error)
                                    return "oldValue." + error;
                            }
                            if (message.updateMask != null && message.hasOwnProperty("updateMask")) {
                                var error = $root.google.events.cloud.firestore.v1.DocumentMask.verify(message.updateMask);
                                if (error)
                                    return "updateMask." + error;
                            }
                            return null;
                        };

                        
                        DocumentEventData.fromObject = function fromObject(object) {
                            if (object instanceof $root.google.events.cloud.firestore.v1.DocumentEventData)
                                return object;
                            var message = new $root.google.events.cloud.firestore.v1.DocumentEventData();
                            if (object.value != null) {
                                if (typeof object.value !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.DocumentEventData.value: object expected");
                                message.value = $root.google.events.cloud.firestore.v1.Document.fromObject(object.value);
                            }
                            if (object.oldValue != null) {
                                if (typeof object.oldValue !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.DocumentEventData.oldValue: object expected");
                                message.oldValue = $root.google.events.cloud.firestore.v1.Document.fromObject(object.oldValue);
                            }
                            if (object.updateMask != null) {
                                if (typeof object.updateMask !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.DocumentEventData.updateMask: object expected");
                                message.updateMask = $root.google.events.cloud.firestore.v1.DocumentMask.fromObject(object.updateMask);
                            }
                            return message;
                        };

                        
                        DocumentEventData.toObject = function toObject(message, options) {
                            if (!options)
                                options = {};
                            var object = {};
                            if (options.defaults) {
                                object.value = null;
                                object.oldValue = null;
                                object.updateMask = null;
                            }
                            if (message.value != null && message.hasOwnProperty("value"))
                                object.value = $root.google.events.cloud.firestore.v1.Document.toObject(message.value, options);
                            if (message.oldValue != null && message.hasOwnProperty("oldValue"))
                                object.oldValue = $root.google.events.cloud.firestore.v1.Document.toObject(message.oldValue, options);
                            if (message.updateMask != null && message.hasOwnProperty("updateMask"))
                                object.updateMask = $root.google.events.cloud.firestore.v1.DocumentMask.toObject(message.updateMask, options);
                            return object;
                        };

                        
                        DocumentEventData.prototype.toJSON = function toJSON() {
                            return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
                        };

                        
                        DocumentEventData.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                            if (typeUrlPrefix === undefined) {
                                typeUrlPrefix = "type.googleapis.com";
                            }
                            return typeUrlPrefix + "/google.events.cloud.firestore.v1.DocumentEventData";
                        };

                        return DocumentEventData;
                    })();

                    v1.DocumentMask = (function() {

                        

                        
                        function DocumentMask(properties) {
                            this.fieldPaths = [];
                            if (properties)
                                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                                    if (properties[keys[i]] != null)
                                        this[keys[i]] = properties[keys[i]];
                        }

                        
                        DocumentMask.prototype.fieldPaths = $util.emptyArray;

                        
                        DocumentMask.create = function create(properties) {
                            return new DocumentMask(properties);
                        };

                        
                        DocumentMask.encode = function encode(message, writer) {
                            if (!writer)
                                writer = $Writer.create();
                            if (message.fieldPaths != null && message.fieldPaths.length)
                                for (var i = 0; i < message.fieldPaths.length; ++i)
                                    writer.uint32(10).string(message.fieldPaths[i]);
                            return writer;
                        };

                        
                        DocumentMask.encodeDelimited = function encodeDelimited(message, writer) {
                            return this.encode(message, writer).ldelim();
                        };

                        
                        DocumentMask.decode = function decode(reader, length) {
                            if (!(reader instanceof $Reader))
                                reader = $Reader.create(reader);
                            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.events.cloud.firestore.v1.DocumentMask();
                            while (reader.pos < end) {
                                var tag = reader.uint32();
                                switch (tag >>> 3) {
                                case 1: {
                                        if (!(message.fieldPaths && message.fieldPaths.length))
                                            message.fieldPaths = [];
                                        message.fieldPaths.push(reader.string());
                                        break;
                                    }
                                default:
                                    reader.skipType(tag & 7);
                                    break;
                                }
                            }
                            return message;
                        };

                        
                        DocumentMask.decodeDelimited = function decodeDelimited(reader) {
                            if (!(reader instanceof $Reader))
                                reader = new $Reader(reader);
                            return this.decode(reader, reader.uint32());
                        };

                        
                        DocumentMask.verify = function verify(message) {
                            if (typeof message !== "object" || message === null)
                                return "object expected";
                            if (message.fieldPaths != null && message.hasOwnProperty("fieldPaths")) {
                                if (!Array.isArray(message.fieldPaths))
                                    return "fieldPaths: array expected";
                                for (var i = 0; i < message.fieldPaths.length; ++i)
                                    if (!$util.isString(message.fieldPaths[i]))
                                        return "fieldPaths: string[] expected";
                            }
                            return null;
                        };

                        
                        DocumentMask.fromObject = function fromObject(object) {
                            if (object instanceof $root.google.events.cloud.firestore.v1.DocumentMask)
                                return object;
                            var message = new $root.google.events.cloud.firestore.v1.DocumentMask();
                            if (object.fieldPaths) {
                                if (!Array.isArray(object.fieldPaths))
                                    throw TypeError(".google.events.cloud.firestore.v1.DocumentMask.fieldPaths: array expected");
                                message.fieldPaths = [];
                                for (var i = 0; i < object.fieldPaths.length; ++i)
                                    message.fieldPaths[i] = String(object.fieldPaths[i]);
                            }
                            return message;
                        };

                        
                        DocumentMask.toObject = function toObject(message, options) {
                            if (!options)
                                options = {};
                            var object = {};
                            if (options.arrays || options.defaults)
                                object.fieldPaths = [];
                            if (message.fieldPaths && message.fieldPaths.length) {
                                object.fieldPaths = [];
                                for (var j = 0; j < message.fieldPaths.length; ++j)
                                    object.fieldPaths[j] = message.fieldPaths[j];
                            }
                            return object;
                        };

                        
                        DocumentMask.prototype.toJSON = function toJSON() {
                            return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
                        };

                        
                        DocumentMask.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                            if (typeUrlPrefix === undefined) {
                                typeUrlPrefix = "type.googleapis.com";
                            }
                            return typeUrlPrefix + "/google.events.cloud.firestore.v1.DocumentMask";
                        };

                        return DocumentMask;
                    })();

                    v1.Document = (function() {

                        

                        
                        function Document(properties) {
                            this.fields = {};
                            if (properties)
                                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                                    if (properties[keys[i]] != null)
                                        this[keys[i]] = properties[keys[i]];
                        }

                        
                        Document.prototype.name = "";

                        
                        Document.prototype.fields = $util.emptyObject;

                        
                        Document.prototype.createTime = null;

                        
                        Document.prototype.updateTime = null;

                        
                        Document.create = function create(properties) {
                            return new Document(properties);
                        };

                        
                        Document.encode = function encode(message, writer) {
                            if (!writer)
                                writer = $Writer.create();
                            if (message.name != null && Object.hasOwnProperty.call(message, "name"))
                                writer.uint32(10).string(message.name);
                            if (message.fields != null && Object.hasOwnProperty.call(message, "fields"))
                                for (var keys = Object.keys(message.fields), i = 0; i < keys.length; ++i) {
                                    writer.uint32(18).fork().uint32(10).string(keys[i]);
                                    $root.google.events.cloud.firestore.v1.Value.encode(message.fields[keys[i]], writer.uint32(18).fork()).ldelim().ldelim();
                                }
                            if (message.createTime != null && Object.hasOwnProperty.call(message, "createTime"))
                                $root.google.protobuf.Timestamp.encode(message.createTime, writer.uint32(26).fork()).ldelim();
                            if (message.updateTime != null && Object.hasOwnProperty.call(message, "updateTime"))
                                $root.google.protobuf.Timestamp.encode(message.updateTime, writer.uint32(34).fork()).ldelim();
                            return writer;
                        };

                        
                        Document.encodeDelimited = function encodeDelimited(message, writer) {
                            return this.encode(message, writer).ldelim();
                        };

                        
                        Document.decode = function decode(reader, length) {
                            if (!(reader instanceof $Reader))
                                reader = $Reader.create(reader);
                            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.events.cloud.firestore.v1.Document(), key, value;
                            while (reader.pos < end) {
                                var tag = reader.uint32();
                                switch (tag >>> 3) {
                                case 1: {
                                        message.name = reader.string();
                                        break;
                                    }
                                case 2: {
                                        if (message.fields === $util.emptyObject)
                                            message.fields = {};
                                        var end2 = reader.uint32() + reader.pos;
                                        key = "";
                                        value = null;
                                        while (reader.pos < end2) {
                                            var tag2 = reader.uint32();
                                            switch (tag2 >>> 3) {
                                            case 1:
                                                key = reader.string();
                                                break;
                                            case 2:
                                                value = $root.google.events.cloud.firestore.v1.Value.decode(reader, reader.uint32());
                                                break;
                                            default:
                                                reader.skipType(tag2 & 7);
                                                break;
                                            }
                                        }
                                        message.fields[key] = value;
                                        break;
                                    }
                                case 3: {
                                        message.createTime = $root.google.protobuf.Timestamp.decode(reader, reader.uint32());
                                        break;
                                    }
                                case 4: {
                                        message.updateTime = $root.google.protobuf.Timestamp.decode(reader, reader.uint32());
                                        break;
                                    }
                                default:
                                    reader.skipType(tag & 7);
                                    break;
                                }
                            }
                            return message;
                        };

                        
                        Document.decodeDelimited = function decodeDelimited(reader) {
                            if (!(reader instanceof $Reader))
                                reader = new $Reader(reader);
                            return this.decode(reader, reader.uint32());
                        };

                        
                        Document.verify = function verify(message) {
                            if (typeof message !== "object" || message === null)
                                return "object expected";
                            if (message.name != null && message.hasOwnProperty("name"))
                                if (!$util.isString(message.name))
                                    return "name: string expected";
                            if (message.fields != null && message.hasOwnProperty("fields")) {
                                if (!$util.isObject(message.fields))
                                    return "fields: object expected";
                                var key = Object.keys(message.fields);
                                for (var i = 0; i < key.length; ++i) {
                                    var error = $root.google.events.cloud.firestore.v1.Value.verify(message.fields[key[i]]);
                                    if (error)
                                        return "fields." + error;
                                }
                            }
                            if (message.createTime != null && message.hasOwnProperty("createTime")) {
                                var error = $root.google.protobuf.Timestamp.verify(message.createTime);
                                if (error)
                                    return "createTime." + error;
                            }
                            if (message.updateTime != null && message.hasOwnProperty("updateTime")) {
                                var error = $root.google.protobuf.Timestamp.verify(message.updateTime);
                                if (error)
                                    return "updateTime." + error;
                            }
                            return null;
                        };

                        
                        Document.fromObject = function fromObject(object) {
                            if (object instanceof $root.google.events.cloud.firestore.v1.Document)
                                return object;
                            var message = new $root.google.events.cloud.firestore.v1.Document();
                            if (object.name != null)
                                message.name = String(object.name);
                            if (object.fields) {
                                if (typeof object.fields !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.Document.fields: object expected");
                                message.fields = {};
                                for (var keys = Object.keys(object.fields), i = 0; i < keys.length; ++i) {
                                    if (typeof object.fields[keys[i]] !== "object")
                                        throw TypeError(".google.events.cloud.firestore.v1.Document.fields: object expected");
                                    message.fields[keys[i]] = $root.google.events.cloud.firestore.v1.Value.fromObject(object.fields[keys[i]]);
                                }
                            }
                            if (object.createTime != null) {
                                if (typeof object.createTime !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.Document.createTime: object expected");
                                message.createTime = $root.google.protobuf.Timestamp.fromObject(object.createTime);
                            }
                            if (object.updateTime != null) {
                                if (typeof object.updateTime !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.Document.updateTime: object expected");
                                message.updateTime = $root.google.protobuf.Timestamp.fromObject(object.updateTime);
                            }
                            return message;
                        };

                        
                        Document.toObject = function toObject(message, options) {
                            if (!options)
                                options = {};
                            var object = {};
                            if (options.objects || options.defaults)
                                object.fields = {};
                            if (options.defaults) {
                                object.name = "";
                                object.createTime = null;
                                object.updateTime = null;
                            }
                            if (message.name != null && message.hasOwnProperty("name"))
                                object.name = message.name;
                            var keys2;
                            if (message.fields && (keys2 = Object.keys(message.fields)).length) {
                                object.fields = {};
                                for (var j = 0; j < keys2.length; ++j)
                                    object.fields[keys2[j]] = $root.google.events.cloud.firestore.v1.Value.toObject(message.fields[keys2[j]], options);
                            }
                            if (message.createTime != null && message.hasOwnProperty("createTime"))
                                object.createTime = $root.google.protobuf.Timestamp.toObject(message.createTime, options);
                            if (message.updateTime != null && message.hasOwnProperty("updateTime"))
                                object.updateTime = $root.google.protobuf.Timestamp.toObject(message.updateTime, options);
                            return object;
                        };

                        
                        Document.prototype.toJSON = function toJSON() {
                            return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
                        };

                        
                        Document.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                            if (typeUrlPrefix === undefined) {
                                typeUrlPrefix = "type.googleapis.com";
                            }
                            return typeUrlPrefix + "/google.events.cloud.firestore.v1.Document";
                        };

                        return Document;
                    })();

                    v1.Value = (function() {

                        

                        
                        function Value(properties) {
                            if (properties)
                                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                                    if (properties[keys[i]] != null)
                                        this[keys[i]] = properties[keys[i]];
                        }

                        
                        Value.prototype.nullValue = null;

                        
                        Value.prototype.booleanValue = null;

                        
                        Value.prototype.integerValue = null;

                        
                        Value.prototype.doubleValue = null;

                        
                        Value.prototype.timestampValue = null;

                        
                        Value.prototype.stringValue = null;

                        
                        Value.prototype.bytesValue = null;

                        
                        Value.prototype.referenceValue = null;

                        
                        Value.prototype.geoPointValue = null;

                        
                        Value.prototype.arrayValue = null;

                        
                        Value.prototype.mapValue = null;


                        var $oneOfFields;

                        
                        Object.defineProperty(Value.prototype, "valueType", {
                            get: $util.oneOfGetter($oneOfFields = ["nullValue", "booleanValue", "integerValue", "doubleValue", "timestampValue", "stringValue", "bytesValue", "referenceValue", "geoPointValue", "arrayValue", "mapValue"]),
                            set: $util.oneOfSetter($oneOfFields)
                        });

                        
                        Value.create = function create(properties) {
                            return new Value(properties);
                        };

                        
                        Value.encode = function encode(message, writer) {
                            if (!writer)
                                writer = $Writer.create();
                            if (message.booleanValue != null && Object.hasOwnProperty.call(message, "booleanValue"))
                                writer.uint32(8).bool(message.booleanValue);
                            if (message.integerValue != null && Object.hasOwnProperty.call(message, "integerValue"))
                                writer.uint32(16).int64(message.integerValue);
                            if (message.doubleValue != null && Object.hasOwnProperty.call(message, "doubleValue"))
                                writer.uint32(25).double(message.doubleValue);
                            if (message.referenceValue != null && Object.hasOwnProperty.call(message, "referenceValue"))
                                writer.uint32(42).string(message.referenceValue);
                            if (message.mapValue != null && Object.hasOwnProperty.call(message, "mapValue"))
                                $root.google.events.cloud.firestore.v1.MapValue.encode(message.mapValue, writer.uint32(50).fork()).ldelim();
                            if (message.geoPointValue != null && Object.hasOwnProperty.call(message, "geoPointValue"))
                                $root.google.type.LatLng.encode(message.geoPointValue, writer.uint32(66).fork()).ldelim();
                            if (message.arrayValue != null && Object.hasOwnProperty.call(message, "arrayValue"))
                                $root.google.events.cloud.firestore.v1.ArrayValue.encode(message.arrayValue, writer.uint32(74).fork()).ldelim();
                            if (message.timestampValue != null && Object.hasOwnProperty.call(message, "timestampValue"))
                                $root.google.protobuf.Timestamp.encode(message.timestampValue, writer.uint32(82).fork()).ldelim();
                            if (message.nullValue != null && Object.hasOwnProperty.call(message, "nullValue"))
                                writer.uint32(88).int32(message.nullValue);
                            if (message.stringValue != null && Object.hasOwnProperty.call(message, "stringValue"))
                                writer.uint32(138).string(message.stringValue);
                            if (message.bytesValue != null && Object.hasOwnProperty.call(message, "bytesValue"))
                                writer.uint32(146).bytes(message.bytesValue);
                            return writer;
                        };

                        
                        Value.encodeDelimited = function encodeDelimited(message, writer) {
                            return this.encode(message, writer).ldelim();
                        };

                        
                        Value.decode = function decode(reader, length) {
                            if (!(reader instanceof $Reader))
                                reader = $Reader.create(reader);
                            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.events.cloud.firestore.v1.Value();
                            while (reader.pos < end) {
                                var tag = reader.uint32();
                                switch (tag >>> 3) {
                                case 11: {
                                        message.nullValue = reader.int32();
                                        break;
                                    }
                                case 1: {
                                        message.booleanValue = reader.bool();
                                        break;
                                    }
                                case 2: {
                                        message.integerValue = reader.int64();
                                        break;
                                    }
                                case 3: {
                                        message.doubleValue = reader.double();
                                        break;
                                    }
                                case 10: {
                                        message.timestampValue = $root.google.protobuf.Timestamp.decode(reader, reader.uint32());
                                        break;
                                    }
                                case 17: {
                                        message.stringValue = reader.string();
                                        break;
                                    }
                                case 18: {
                                        message.bytesValue = reader.bytes();
                                        break;
                                    }
                                case 5: {
                                        message.referenceValue = reader.string();
                                        break;
                                    }
                                case 8: {
                                        message.geoPointValue = $root.google.type.LatLng.decode(reader, reader.uint32());
                                        break;
                                    }
                                case 9: {
                                        message.arrayValue = $root.google.events.cloud.firestore.v1.ArrayValue.decode(reader, reader.uint32());
                                        break;
                                    }
                                case 6: {
                                        message.mapValue = $root.google.events.cloud.firestore.v1.MapValue.decode(reader, reader.uint32());
                                        break;
                                    }
                                default:
                                    reader.skipType(tag & 7);
                                    break;
                                }
                            }
                            return message;
                        };

                        
                        Value.decodeDelimited = function decodeDelimited(reader) {
                            if (!(reader instanceof $Reader))
                                reader = new $Reader(reader);
                            return this.decode(reader, reader.uint32());
                        };

                        
                        Value.verify = function verify(message) {
                            if (typeof message !== "object" || message === null)
                                return "object expected";
                            var properties = {};
                            if (message.nullValue != null && message.hasOwnProperty("nullValue")) {
                                properties.valueType = 1;
                                switch (message.nullValue) {
                                default:
                                    return "nullValue: enum value expected";
                                case 0:
                                    break;
                                }
                            }
                            if (message.booleanValue != null && message.hasOwnProperty("booleanValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                if (typeof message.booleanValue !== "boolean")
                                    return "booleanValue: boolean expected";
                            }
                            if (message.integerValue != null && message.hasOwnProperty("integerValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                if (!$util.isInteger(message.integerValue) && !(message.integerValue && $util.isInteger(message.integerValue.low) && $util.isInteger(message.integerValue.high)))
                                    return "integerValue: integer|Long expected";
                            }
                            if (message.doubleValue != null && message.hasOwnProperty("doubleValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                if (typeof message.doubleValue !== "number")
                                    return "doubleValue: number expected";
                            }
                            if (message.timestampValue != null && message.hasOwnProperty("timestampValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                {
                                    var error = $root.google.protobuf.Timestamp.verify(message.timestampValue);
                                    if (error)
                                        return "timestampValue." + error;
                                }
                            }
                            if (message.stringValue != null && message.hasOwnProperty("stringValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                if (!$util.isString(message.stringValue))
                                    return "stringValue: string expected";
                            }
                            if (message.bytesValue != null && message.hasOwnProperty("bytesValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                if (!(message.bytesValue && typeof message.bytesValue.length === "number" || $util.isString(message.bytesValue)))
                                    return "bytesValue: buffer expected";
                            }
                            if (message.referenceValue != null && message.hasOwnProperty("referenceValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                if (!$util.isString(message.referenceValue))
                                    return "referenceValue: string expected";
                            }
                            if (message.geoPointValue != null && message.hasOwnProperty("geoPointValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                {
                                    var error = $root.google.type.LatLng.verify(message.geoPointValue);
                                    if (error)
                                        return "geoPointValue." + error;
                                }
                            }
                            if (message.arrayValue != null && message.hasOwnProperty("arrayValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                {
                                    var error = $root.google.events.cloud.firestore.v1.ArrayValue.verify(message.arrayValue);
                                    if (error)
                                        return "arrayValue." + error;
                                }
                            }
                            if (message.mapValue != null && message.hasOwnProperty("mapValue")) {
                                if (properties.valueType === 1)
                                    return "valueType: multiple values";
                                properties.valueType = 1;
                                {
                                    var error = $root.google.events.cloud.firestore.v1.MapValue.verify(message.mapValue);
                                    if (error)
                                        return "mapValue." + error;
                                }
                            }
                            return null;
                        };

                        
                        Value.fromObject = function fromObject(object) {
                            if (object instanceof $root.google.events.cloud.firestore.v1.Value)
                                return object;
                            var message = new $root.google.events.cloud.firestore.v1.Value();
                            switch (object.nullValue) {
                            default:
                                if (typeof object.nullValue === "number") {
                                    message.nullValue = object.nullValue;
                                    break;
                                }
                                break;
                            case "NULL_VALUE":
                            case 0:
                                message.nullValue = 0;
                                break;
                            }
                            if (object.booleanValue != null)
                                message.booleanValue = Boolean(object.booleanValue);
                            if (object.integerValue != null)
                                if ($util.Long)
                                    (message.integerValue = $util.Long.fromValue(object.integerValue)).unsigned = false;
                                else if (typeof object.integerValue === "string")
                                    message.integerValue = parseInt(object.integerValue, 10);
                                else if (typeof object.integerValue === "number")
                                    message.integerValue = object.integerValue;
                                else if (typeof object.integerValue === "object")
                                    message.integerValue = new $util.LongBits(object.integerValue.low >>> 0, object.integerValue.high >>> 0).toNumber();
                            if (object.doubleValue != null)
                                message.doubleValue = Number(object.doubleValue);
                            if (object.timestampValue != null) {
                                if (typeof object.timestampValue !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.Value.timestampValue: object expected");
                                message.timestampValue = $root.google.protobuf.Timestamp.fromObject(object.timestampValue);
                            }
                            if (object.stringValue != null)
                                message.stringValue = String(object.stringValue);
                            if (object.bytesValue != null)
                                if (typeof object.bytesValue === "string")
                                    $util.base64.decode(object.bytesValue, message.bytesValue = $util.newBuffer($util.base64.length(object.bytesValue)), 0);
                                else if (object.bytesValue.length >= 0)
                                    message.bytesValue = object.bytesValue;
                            if (object.referenceValue != null)
                                message.referenceValue = String(object.referenceValue);
                            if (object.geoPointValue != null) {
                                if (typeof object.geoPointValue !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.Value.geoPointValue: object expected");
                                message.geoPointValue = $root.google.type.LatLng.fromObject(object.geoPointValue);
                            }
                            if (object.arrayValue != null) {
                                if (typeof object.arrayValue !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.Value.arrayValue: object expected");
                                message.arrayValue = $root.google.events.cloud.firestore.v1.ArrayValue.fromObject(object.arrayValue);
                            }
                            if (object.mapValue != null) {
                                if (typeof object.mapValue !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.Value.mapValue: object expected");
                                message.mapValue = $root.google.events.cloud.firestore.v1.MapValue.fromObject(object.mapValue);
                            }
                            return message;
                        };

                        
                        Value.toObject = function toObject(message, options) {
                            if (!options)
                                options = {};
                            var object = {};
                            if (message.booleanValue != null && message.hasOwnProperty("booleanValue")) {
                                object.booleanValue = message.booleanValue;
                                if (options.oneofs)
                                    object.valueType = "booleanValue";
                            }
                            if (message.integerValue != null && message.hasOwnProperty("integerValue")) {
                                if (typeof message.integerValue === "number")
                                    object.integerValue = options.longs === String ? String(message.integerValue) : message.integerValue;
                                else
                                    object.integerValue = options.longs === String ? $util.Long.prototype.toString.call(message.integerValue) : options.longs === Number ? new $util.LongBits(message.integerValue.low >>> 0, message.integerValue.high >>> 0).toNumber() : message.integerValue;
                                if (options.oneofs)
                                    object.valueType = "integerValue";
                            }
                            if (message.doubleValue != null && message.hasOwnProperty("doubleValue")) {
                                object.doubleValue = options.json && !isFinite(message.doubleValue) ? String(message.doubleValue) : message.doubleValue;
                                if (options.oneofs)
                                    object.valueType = "doubleValue";
                            }
                            if (message.referenceValue != null && message.hasOwnProperty("referenceValue")) {
                                object.referenceValue = message.referenceValue;
                                if (options.oneofs)
                                    object.valueType = "referenceValue";
                            }
                            if (message.mapValue != null && message.hasOwnProperty("mapValue")) {
                                object.mapValue = $root.google.events.cloud.firestore.v1.MapValue.toObject(message.mapValue, options);
                                if (options.oneofs)
                                    object.valueType = "mapValue";
                            }
                            if (message.geoPointValue != null && message.hasOwnProperty("geoPointValue")) {
                                object.geoPointValue = $root.google.type.LatLng.toObject(message.geoPointValue, options);
                                if (options.oneofs)
                                    object.valueType = "geoPointValue";
                            }
                            if (message.arrayValue != null && message.hasOwnProperty("arrayValue")) {
                                object.arrayValue = $root.google.events.cloud.firestore.v1.ArrayValue.toObject(message.arrayValue, options);
                                if (options.oneofs)
                                    object.valueType = "arrayValue";
                            }
                            if (message.timestampValue != null && message.hasOwnProperty("timestampValue")) {
                                object.timestampValue = $root.google.protobuf.Timestamp.toObject(message.timestampValue, options);
                                if (options.oneofs)
                                    object.valueType = "timestampValue";
                            }
                            if (message.nullValue != null && message.hasOwnProperty("nullValue")) {
                                object.nullValue = options.enums === String ? $root.google.protobuf.NullValue[message.nullValue] === undefined ? message.nullValue : $root.google.protobuf.NullValue[message.nullValue] : message.nullValue;
                                if (options.oneofs)
                                    object.valueType = "nullValue";
                            }
                            if (message.stringValue != null && message.hasOwnProperty("stringValue")) {
                                object.stringValue = message.stringValue;
                                if (options.oneofs)
                                    object.valueType = "stringValue";
                            }
                            if (message.bytesValue != null && message.hasOwnProperty("bytesValue")) {
                                object.bytesValue = options.bytes === String ? $util.base64.encode(message.bytesValue, 0, message.bytesValue.length) : options.bytes === Array ? Array.prototype.slice.call(message.bytesValue) : message.bytesValue;
                                if (options.oneofs)
                                    object.valueType = "bytesValue";
                            }
                            return object;
                        };

                        
                        Value.prototype.toJSON = function toJSON() {
                            return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
                        };

                        
                        Value.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                            if (typeUrlPrefix === undefined) {
                                typeUrlPrefix = "type.googleapis.com";
                            }
                            return typeUrlPrefix + "/google.events.cloud.firestore.v1.Value";
                        };

                        return Value;
                    })();

                    v1.ArrayValue = (function() {

                        

                        
                        function ArrayValue(properties) {
                            this.values = [];
                            if (properties)
                                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                                    if (properties[keys[i]] != null)
                                        this[keys[i]] = properties[keys[i]];
                        }

                        
                        ArrayValue.prototype.values = $util.emptyArray;

                        
                        ArrayValue.create = function create(properties) {
                            return new ArrayValue(properties);
                        };

                        
                        ArrayValue.encode = function encode(message, writer) {
                            if (!writer)
                                writer = $Writer.create();
                            if (message.values != null && message.values.length)
                                for (var i = 0; i < message.values.length; ++i)
                                    $root.google.events.cloud.firestore.v1.Value.encode(message.values[i], writer.uint32(10).fork()).ldelim();
                            return writer;
                        };

                        
                        ArrayValue.encodeDelimited = function encodeDelimited(message, writer) {
                            return this.encode(message, writer).ldelim();
                        };

                        
                        ArrayValue.decode = function decode(reader, length) {
                            if (!(reader instanceof $Reader))
                                reader = $Reader.create(reader);
                            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.events.cloud.firestore.v1.ArrayValue();
                            while (reader.pos < end) {
                                var tag = reader.uint32();
                                switch (tag >>> 3) {
                                case 1: {
                                        if (!(message.values && message.values.length))
                                            message.values = [];
                                        message.values.push($root.google.events.cloud.firestore.v1.Value.decode(reader, reader.uint32()));
                                        break;
                                    }
                                default:
                                    reader.skipType(tag & 7);
                                    break;
                                }
                            }
                            return message;
                        };

                        
                        ArrayValue.decodeDelimited = function decodeDelimited(reader) {
                            if (!(reader instanceof $Reader))
                                reader = new $Reader(reader);
                            return this.decode(reader, reader.uint32());
                        };

                        
                        ArrayValue.verify = function verify(message) {
                            if (typeof message !== "object" || message === null)
                                return "object expected";
                            if (message.values != null && message.hasOwnProperty("values")) {
                                if (!Array.isArray(message.values))
                                    return "values: array expected";
                                for (var i = 0; i < message.values.length; ++i) {
                                    var error = $root.google.events.cloud.firestore.v1.Value.verify(message.values[i]);
                                    if (error)
                                        return "values." + error;
                                }
                            }
                            return null;
                        };

                        
                        ArrayValue.fromObject = function fromObject(object) {
                            if (object instanceof $root.google.events.cloud.firestore.v1.ArrayValue)
                                return object;
                            var message = new $root.google.events.cloud.firestore.v1.ArrayValue();
                            if (object.values) {
                                if (!Array.isArray(object.values))
                                    throw TypeError(".google.events.cloud.firestore.v1.ArrayValue.values: array expected");
                                message.values = [];
                                for (var i = 0; i < object.values.length; ++i) {
                                    if (typeof object.values[i] !== "object")
                                        throw TypeError(".google.events.cloud.firestore.v1.ArrayValue.values: object expected");
                                    message.values[i] = $root.google.events.cloud.firestore.v1.Value.fromObject(object.values[i]);
                                }
                            }
                            return message;
                        };

                        
                        ArrayValue.toObject = function toObject(message, options) {
                            if (!options)
                                options = {};
                            var object = {};
                            if (options.arrays || options.defaults)
                                object.values = [];
                            if (message.values && message.values.length) {
                                object.values = [];
                                for (var j = 0; j < message.values.length; ++j)
                                    object.values[j] = $root.google.events.cloud.firestore.v1.Value.toObject(message.values[j], options);
                            }
                            return object;
                        };

                        
                        ArrayValue.prototype.toJSON = function toJSON() {
                            return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
                        };

                        
                        ArrayValue.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                            if (typeUrlPrefix === undefined) {
                                typeUrlPrefix = "type.googleapis.com";
                            }
                            return typeUrlPrefix + "/google.events.cloud.firestore.v1.ArrayValue";
                        };

                        return ArrayValue;
                    })();

                    v1.MapValue = (function() {

                        

                        
                        function MapValue(properties) {
                            this.fields = {};
                            if (properties)
                                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                                    if (properties[keys[i]] != null)
                                        this[keys[i]] = properties[keys[i]];
                        }

                        
                        MapValue.prototype.fields = $util.emptyObject;

                        
                        MapValue.create = function create(properties) {
                            return new MapValue(properties);
                        };

                        
                        MapValue.encode = function encode(message, writer) {
                            if (!writer)
                                writer = $Writer.create();
                            if (message.fields != null && Object.hasOwnProperty.call(message, "fields"))
                                for (var keys = Object.keys(message.fields), i = 0; i < keys.length; ++i) {
                                    writer.uint32(10).fork().uint32(10).string(keys[i]);
                                    $root.google.events.cloud.firestore.v1.Value.encode(message.fields[keys[i]], writer.uint32(18).fork()).ldelim().ldelim();
                                }
                            return writer;
                        };

                        
                        MapValue.encodeDelimited = function encodeDelimited(message, writer) {
                            return this.encode(message, writer).ldelim();
                        };

                        
                        MapValue.decode = function decode(reader, length) {
                            if (!(reader instanceof $Reader))
                                reader = $Reader.create(reader);
                            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.events.cloud.firestore.v1.MapValue(), key, value;
                            while (reader.pos < end) {
                                var tag = reader.uint32();
                                switch (tag >>> 3) {
                                case 1: {
                                        if (message.fields === $util.emptyObject)
                                            message.fields = {};
                                        var end2 = reader.uint32() + reader.pos;
                                        key = "";
                                        value = null;
                                        while (reader.pos < end2) {
                                            var tag2 = reader.uint32();
                                            switch (tag2 >>> 3) {
                                            case 1:
                                                key = reader.string();
                                                break;
                                            case 2:
                                                value = $root.google.events.cloud.firestore.v1.Value.decode(reader, reader.uint32());
                                                break;
                                            default:
                                                reader.skipType(tag2 & 7);
                                                break;
                                            }
                                        }
                                        message.fields[key] = value;
                                        break;
                                    }
                                default:
                                    reader.skipType(tag & 7);
                                    break;
                                }
                            }
                            return message;
                        };

                        
                        MapValue.decodeDelimited = function decodeDelimited(reader) {
                            if (!(reader instanceof $Reader))
                                reader = new $Reader(reader);
                            return this.decode(reader, reader.uint32());
                        };

                        
                        MapValue.verify = function verify(message) {
                            if (typeof message !== "object" || message === null)
                                return "object expected";
                            if (message.fields != null && message.hasOwnProperty("fields")) {
                                if (!$util.isObject(message.fields))
                                    return "fields: object expected";
                                var key = Object.keys(message.fields);
                                for (var i = 0; i < key.length; ++i) {
                                    var error = $root.google.events.cloud.firestore.v1.Value.verify(message.fields[key[i]]);
                                    if (error)
                                        return "fields." + error;
                                }
                            }
                            return null;
                        };

                        
                        MapValue.fromObject = function fromObject(object) {
                            if (object instanceof $root.google.events.cloud.firestore.v1.MapValue)
                                return object;
                            var message = new $root.google.events.cloud.firestore.v1.MapValue();
                            if (object.fields) {
                                if (typeof object.fields !== "object")
                                    throw TypeError(".google.events.cloud.firestore.v1.MapValue.fields: object expected");
                                message.fields = {};
                                for (var keys = Object.keys(object.fields), i = 0; i < keys.length; ++i) {
                                    if (typeof object.fields[keys[i]] !== "object")
                                        throw TypeError(".google.events.cloud.firestore.v1.MapValue.fields: object expected");
                                    message.fields[keys[i]] = $root.google.events.cloud.firestore.v1.Value.fromObject(object.fields[keys[i]]);
                                }
                            }
                            return message;
                        };

                        
                        MapValue.toObject = function toObject(message, options) {
                            if (!options)
                                options = {};
                            var object = {};
                            if (options.objects || options.defaults)
                                object.fields = {};
                            var keys2;
                            if (message.fields && (keys2 = Object.keys(message.fields)).length) {
                                object.fields = {};
                                for (var j = 0; j < keys2.length; ++j)
                                    object.fields[keys2[j]] = $root.google.events.cloud.firestore.v1.Value.toObject(message.fields[keys2[j]], options);
                            }
                            return object;
                        };

                        
                        MapValue.prototype.toJSON = function toJSON() {
                            return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
                        };

                        
                        MapValue.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                            if (typeUrlPrefix === undefined) {
                                typeUrlPrefix = "type.googleapis.com";
                            }
                            return typeUrlPrefix + "/google.events.cloud.firestore.v1.MapValue";
                        };

                        return MapValue;
                    })();

                    return v1;
                })();

                return firestore;
            })();

            return cloud;
        })();

        return events;
    })();

    google.type = (function() {

        
        var type = {};

        type.LatLng = (function() {

            

            
            function LatLng(properties) {
                if (properties)
                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                        if (properties[keys[i]] != null)
                            this[keys[i]] = properties[keys[i]];
            }

            
            LatLng.prototype.latitude = 0;

            
            LatLng.prototype.longitude = 0;

            
            LatLng.create = function create(properties) {
                return new LatLng(properties);
            };

            
            LatLng.encode = function encode(message, writer) {
                if (!writer)
                    writer = $Writer.create();
                if (message.latitude != null && Object.hasOwnProperty.call(message, "latitude"))
                    writer.uint32(9).double(message.latitude);
                if (message.longitude != null && Object.hasOwnProperty.call(message, "longitude"))
                    writer.uint32(17).double(message.longitude);
                return writer;
            };

            
            LatLng.encodeDelimited = function encodeDelimited(message, writer) {
                return this.encode(message, writer).ldelim();
            };

            
            LatLng.decode = function decode(reader, length) {
                if (!(reader instanceof $Reader))
                    reader = $Reader.create(reader);
                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.type.LatLng();
                while (reader.pos < end) {
                    var tag = reader.uint32();
                    switch (tag >>> 3) {
                    case 1: {
                            message.latitude = reader.double();
                            break;
                        }
                    case 2: {
                            message.longitude = reader.double();
                            break;
                        }
                    default:
                        reader.skipType(tag & 7);
                        break;
                    }
                }
                return message;
            };

            
            LatLng.decodeDelimited = function decodeDelimited(reader) {
                if (!(reader instanceof $Reader))
                    reader = new $Reader(reader);
                return this.decode(reader, reader.uint32());
            };

            
            LatLng.verify = function verify(message) {
                if (typeof message !== "object" || message === null)
                    return "object expected";
                if (message.latitude != null && message.hasOwnProperty("latitude"))
                    if (typeof message.latitude !== "number")
                        return "latitude: number expected";
                if (message.longitude != null && message.hasOwnProperty("longitude"))
                    if (typeof message.longitude !== "number")
                        return "longitude: number expected";
                return null;
            };

            
            LatLng.fromObject = function fromObject(object) {
                if (object instanceof $root.google.type.LatLng)
                    return object;
                var message = new $root.google.type.LatLng();
                if (object.latitude != null)
                    message.latitude = Number(object.latitude);
                if (object.longitude != null)
                    message.longitude = Number(object.longitude);
                return message;
            };

            
            LatLng.toObject = function toObject(message, options) {
                if (!options)
                    options = {};
                var object = {};
                if (options.defaults) {
                    object.latitude = 0;
                    object.longitude = 0;
                }
                if (message.latitude != null && message.hasOwnProperty("latitude"))
                    object.latitude = options.json && !isFinite(message.latitude) ? String(message.latitude) : message.latitude;
                if (message.longitude != null && message.hasOwnProperty("longitude"))
                    object.longitude = options.json && !isFinite(message.longitude) ? String(message.longitude) : message.longitude;
                return object;
            };

            
            LatLng.prototype.toJSON = function toJSON() {
                return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
            };

            
            LatLng.getTypeUrl = function getTypeUrl(typeUrlPrefix) {
                if (typeUrlPrefix === undefined) {
                    typeUrlPrefix = "type.googleapis.com";
                }
                return typeUrlPrefix + "/google.type.LatLng";
            };

            return LatLng;
        })();

        return type;
    })();

    return google;
})();

module.exports = $root;
