// firebase-functions/lib/v2/providers/alerts/appDistribution.js
const require_rolldown_runtime = require('../../../_virtual/rolldown_runtime.js');
const require_common_onInit = require('../../../common/onInit.js');
const require_v2_trace = require('../../trace.js');
const require_v2_providers_alerts_alerts = require('./alerts.js');


var appDistribution_exports =  require_rolldown_runtime.__export({
	getOptsAndApp: () => getOptsAndApp,
	inAppFeedbackAlert: () => inAppFeedbackAlert,
	newTesterIosDeviceAlert: () => newTesterIosDeviceAlert,
	onInAppFeedbackPublished: () => onInAppFeedbackPublished,
	onNewTesterIosDevicePublished: () => onNewTesterIosDevicePublished
});

const newTesterIosDeviceAlert = "appDistribution.newTesterIosDevice";

const inAppFeedbackAlert = "appDistribution.inAppFeedback";

function onNewTesterIosDevicePublished(appIdOrOptsOrHandler, handler) {
	if (typeof appIdOrOptsOrHandler === "function") {
		handler = appIdOrOptsOrHandler;
		appIdOrOptsOrHandler = {};
	}
	const [opts, appId] = getOptsAndApp(appIdOrOptsOrHandler);
	const func = (raw) => {
		return require_v2_trace.wrapTraceContext(require_common_onInit.withInit(handler))(require_v2_providers_alerts_alerts.convertAlertAndApp(raw));
	};
	func.run = handler;
	func.__endpoint = require_v2_providers_alerts_alerts.getEndpointAnnotation(opts, newTesterIosDeviceAlert, appId);
	return func;
}

function onInAppFeedbackPublished(appIdOrOptsOrHandler, handler) {
	if (typeof appIdOrOptsOrHandler === "function") {
		handler = appIdOrOptsOrHandler;
		appIdOrOptsOrHandler = {};
	}
	const [opts, appId] = getOptsAndApp(appIdOrOptsOrHandler);
	const func = (raw) => {
		return require_v2_trace.wrapTraceContext(require_common_onInit.withInit(handler))(require_v2_providers_alerts_alerts.convertAlertAndApp(raw));
	};
	func.run = handler;
	func.__endpoint = require_v2_providers_alerts_alerts.getEndpointAnnotation(opts, inAppFeedbackAlert, appId);
	return func;
}

function getOptsAndApp(appIdOrOpts) {
	let opts;
	let appId;
	if (typeof appIdOrOpts === "string") {
		opts = {};
		appId = appIdOrOpts;
	} else {
		appId = appIdOrOpts.appId;
		opts = { ...appIdOrOpts };
		delete opts.appId;
	}
	return [opts, appId];
}


Object.defineProperty(exports, 'appDistribution_exports', {
  enumerable: true,
  get: function () {
    return appDistribution_exports;
  }
});
exports.getOptsAndApp = getOptsAndApp;
exports.inAppFeedbackAlert = inAppFeedbackAlert;
exports.newTesterIosDeviceAlert = newTesterIosDeviceAlert;
exports.onInAppFeedbackPublished = onInAppFeedbackPublished;
exports.onNewTesterIosDevicePublished = onNewTesterIosDevicePublished;
