// firebase-functions/lib/v2/providers/alerts/billing.js
const require_rolldown_runtime = require('../../../_virtual/rolldown_runtime.js');
const require_common_onInit = require('../../../common/onInit.js');
const require_v2_trace = require('../../trace.js');
const require_v2_providers_alerts_alerts = require('./alerts.js');


var billing_exports =  require_rolldown_runtime.__export({
	onOperation: () => onOperation,
	onPlanAutomatedUpdatePublished: () => onPlanAutomatedUpdatePublished,
	onPlanUpdatePublished: () => onPlanUpdatePublished,
	planAutomatedUpdateAlert: () => planAutomatedUpdateAlert,
	planUpdateAlert: () => planUpdateAlert
});

const planUpdateAlert = "billing.planUpdate";

const planAutomatedUpdateAlert = "billing.planAutomatedUpdate";

function onPlanUpdatePublished(optsOrHandler, handler) {
	return onOperation(planUpdateAlert, optsOrHandler, handler);
}

function onPlanAutomatedUpdatePublished(optsOrHandler, handler) {
	return onOperation(planAutomatedUpdateAlert, optsOrHandler, handler);
}

function onOperation(alertType, optsOrHandler, handler) {
	if (typeof optsOrHandler === "function") {
		handler = optsOrHandler;
		optsOrHandler = {};
	}
	const func = (raw) => {
		return require_v2_trace.wrapTraceContext(require_common_onInit.withInit(handler))(require_v2_providers_alerts_alerts.convertAlertAndApp(raw));
	};
	func.run = handler;
	func.__endpoint = require_v2_providers_alerts_alerts.getEndpointAnnotation(optsOrHandler, alertType);
	return func;
}


Object.defineProperty(exports, 'billing_exports', {
  enumerable: true,
  get: function () {
    return billing_exports;
  }
});
exports.onOperation = onOperation;
exports.onPlanAutomatedUpdatePublished = onPlanAutomatedUpdatePublished;
exports.onPlanUpdatePublished = onPlanUpdatePublished;
exports.planAutomatedUpdateAlert = planAutomatedUpdateAlert;
exports.planUpdateAlert = planUpdateAlert;
