// firebase-functions/lib/v2/providers/alerts/crashlytics.js
const require_rolldown_runtime = require('../../../_virtual/rolldown_runtime.js');
const require_common_onInit = require('../../../common/onInit.js');
const require_v2_trace = require('../../trace.js');
const require_v2_providers_alerts_alerts = require('./alerts.js');


var crashlytics_exports =  require_rolldown_runtime.__export({
	getOptsAndApp: () => getOptsAndApp,
	newAnrIssueAlert: () => newAnrIssueAlert,
	newFatalIssueAlert: () => newFatalIssueAlert,
	newNonfatalIssueAlert: () => newNonfatalIssueAlert,
	onNewAnrIssuePublished: () => onNewAnrIssuePublished,
	onNewFatalIssuePublished: () => onNewFatalIssuePublished,
	onNewNonfatalIssuePublished: () => onNewNonfatalIssuePublished,
	onOperation: () => onOperation,
	onRegressionAlertPublished: () => onRegressionAlertPublished,
	onStabilityDigestPublished: () => onStabilityDigestPublished,
	onVelocityAlertPublished: () => onVelocityAlertPublished,
	regressionAlert: () => regressionAlert,
	stabilityDigestAlert: () => stabilityDigestAlert,
	velocityAlert: () => velocityAlert
});

const newFatalIssueAlert = "crashlytics.newFatalIssue";

const newNonfatalIssueAlert = "crashlytics.newNonfatalIssue";

const regressionAlert = "crashlytics.regression";

const stabilityDigestAlert = "crashlytics.stabilityDigest";

const velocityAlert = "crashlytics.velocity";

const newAnrIssueAlert = "crashlytics.newAnrIssue";

function onNewFatalIssuePublished(appIdOrOptsOrHandler, handler) {
	return onOperation(newFatalIssueAlert, appIdOrOptsOrHandler, handler);
}

function onNewNonfatalIssuePublished(appIdOrOptsOrHandler, handler) {
	return onOperation(newNonfatalIssueAlert, appIdOrOptsOrHandler, handler);
}

function onRegressionAlertPublished(appIdOrOptsOrHandler, handler) {
	return onOperation(regressionAlert, appIdOrOptsOrHandler, handler);
}

function onStabilityDigestPublished(appIdOrOptsOrHandler, handler) {
	return onOperation(stabilityDigestAlert, appIdOrOptsOrHandler, handler);
}

function onVelocityAlertPublished(appIdOrOptsOrHandler, handler) {
	return onOperation(velocityAlert, appIdOrOptsOrHandler, handler);
}

function onNewAnrIssuePublished(appIdOrOptsOrHandler, handler) {
	return onOperation(newAnrIssueAlert, appIdOrOptsOrHandler, handler);
}

function onOperation(alertType, appIdOrOptsOrHandler, handler) {
	if (typeof appIdOrOptsOrHandler === "function") {
		handler = appIdOrOptsOrHandler;
		appIdOrOptsOrHandler = {};
	}
	const [opts, appId] = getOptsAndApp(appIdOrOptsOrHandler);
	const func = (raw) => {
		return require_v2_trace.wrapTraceContext(require_common_onInit.withInit(handler))(require_v2_providers_alerts_alerts.convertAlertAndApp(raw));
	};
	func.run = handler;
	func.__endpoint = require_v2_providers_alerts_alerts.getEndpointAnnotation(opts, alertType, appId);
	return func;
}

function getOptsAndApp(appIdOrOpts) {
	let opts;
	let appId;
	if (typeof appIdOrOpts === "string") {
		opts = {};
		appId = appIdOrOpts;
	} else {
		appId = appIdOrOpts.appId;
		opts = { ...appIdOrOpts };
		delete opts.appId;
	}
	return [opts, appId];
}


Object.defineProperty(exports, 'crashlytics_exports', {
  enumerable: true,
  get: function () {
    return crashlytics_exports;
  }
});
exports.getOptsAndApp = getOptsAndApp;
exports.newAnrIssueAlert = newAnrIssueAlert;
exports.newFatalIssueAlert = newFatalIssueAlert;
exports.newNonfatalIssueAlert = newNonfatalIssueAlert;
exports.onNewAnrIssuePublished = onNewAnrIssuePublished;
exports.onNewFatalIssuePublished = onNewFatalIssuePublished;
exports.onNewNonfatalIssuePublished = onNewNonfatalIssuePublished;
exports.onOperation = onOperation;
exports.onRegressionAlertPublished = onRegressionAlertPublished;
exports.onStabilityDigestPublished = onStabilityDigestPublished;
exports.onVelocityAlertPublished = onVelocityAlertPublished;
exports.regressionAlert = regressionAlert;
exports.stabilityDigestAlert = stabilityDigestAlert;
exports.velocityAlert = velocityAlert;
