// firebase-functions/lib/v1/providers/database.js
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.js');
const require_common_config = require('../../common/config.js');
const require_common_app = require('../../common/app.js');
const require_v1_cloud_functions = require('../cloud-functions.js');
const require_common_utilities_path = require('../../common/utilities/path.js');
const require_common_providers_database = require('../../common/providers/database.js');
const require_common_utilities_utils = require('../../common/utilities/utils.js');


var database_exports =  require_rolldown_runtime.__export({
	DataSnapshot: () => require_common_providers_database.DataSnapshot,
	InstanceBuilder: () => InstanceBuilder,
	RefBuilder: () => RefBuilder,
	_instanceWithOptions: () => _instanceWithOptions,
	_refWithOptions: () => _refWithOptions,
	extractInstanceAndPath: () => extractInstanceAndPath,
	instance: () => instance,
	provider: () => provider,
	ref: () => ref,
	service: () => service
});

const provider = "google.firebase.database";

const service = "firebaseio.com";
const databaseURLRegex = new RegExp("^https://([^.]+).");
const emulatorDatabaseURLRegex = new RegExp("^http://.*ns=([^&]+)");

function instance(instance$1) {
	return _instanceWithOptions(instance$1, {});
}

function ref(path) {
	return _refWithOptions(path, {});
}

function _instanceWithOptions(instance$1, options) {
	return new InstanceBuilder(instance$1, options);
}

var InstanceBuilder = class {
	constructor(instance$1, options) {
		this.instance = instance$1;
		this.options = options;
	}
	
	ref(path) {
		const normalized = require_common_utilities_path.normalizePath(path);
		return new RefBuilder(() => `projects/_/instances/${this.instance}/refs/${normalized}`, this.options);
	}
};

function _refWithOptions(path, options) {
	const resourceGetter = () => {
		const normalized = require_common_utilities_path.normalizePath(path);
		const databaseURL = require_common_config.firebaseConfig().databaseURL;
		if (!databaseURL) {
			throw new Error("Missing expected firebase config value databaseURL, " + "config is actually" + JSON.stringify(require_common_config.firebaseConfig()) + "\n If you are unit testing, please set process.env.FIREBASE_CONFIG");
		}
		let instance$1;
		const prodMatch = databaseURL.match(databaseURLRegex);
		if (prodMatch) {
			instance$1 = prodMatch[1];
		} else {
			const emulatorMatch = databaseURL.match(emulatorDatabaseURLRegex);
			if (emulatorMatch) {
				instance$1 = emulatorMatch[1];
			}
		}
		if (!instance$1) {
			throw new Error("Invalid value for config firebase.databaseURL: " + databaseURL);
		}
		return `projects/_/instances/${instance$1}/refs/${normalized}`;
	};
	return new RefBuilder(resourceGetter, options);
}

var RefBuilder = class {
	constructor(triggerResource, options) {
		this.triggerResource = triggerResource;
		this.options = options;
		this.changeConstructor = (raw) => {
			const [dbInstance, path] = extractInstanceAndPath(raw.context.resource.name, raw.context.domain);
			const before = new require_common_providers_database.DataSnapshot(raw.data.data, path, require_common_app.getApp(), dbInstance);
			const after = new require_common_providers_database.DataSnapshot(require_common_utilities_utils.applyChange(raw.data.data, raw.data.delta), path, require_common_app.getApp(), dbInstance);
			return {
				before,
				after
			};
		};
	}
	
	onWrite(handler) {
		return this.onOperation(handler, "ref.write", this.changeConstructor);
	}
	
	onUpdate(handler) {
		return this.onOperation(handler, "ref.update", this.changeConstructor);
	}
	
	onCreate(handler) {
		const dataConstructor = (raw) => {
			const [dbInstance, path] = extractInstanceAndPath(raw.context.resource.name, raw.context.domain);
			return new require_common_providers_database.DataSnapshot(raw.data.delta, path, require_common_app.getApp(), dbInstance);
		};
		return this.onOperation(handler, "ref.create", dataConstructor);
	}
	
	onDelete(handler) {
		const dataConstructor = (raw) => {
			const [dbInstance, path] = extractInstanceAndPath(raw.context.resource.name, raw.context.domain);
			return new require_common_providers_database.DataSnapshot(raw.data.data, path, require_common_app.getApp(), dbInstance);
		};
		return this.onOperation(handler, "ref.delete", dataConstructor);
	}
	onOperation(handler, eventType, dataConstructor) {
		return require_v1_cloud_functions.makeCloudFunction({
			handler,
			provider,
			service,
			eventType,
			legacyEventType: `providers/${provider}/eventTypes/${eventType}`,
			triggerResource: this.triggerResource,
			dataConstructor,
			options: this.options
		});
	}
};
const resourceRegex = /^projects\/([^/]+)\/instances\/([a-zA-Z0-9-]+)\/refs(\/.+)?/;

function extractInstanceAndPath(resource, domain = "firebaseio.com") {
	const match = resource.match(new RegExp(resourceRegex));
	if (!match) {
		throw new Error(`Unexpected resource string for Firebase Realtime Database event: ${resource}. ` + "Expected string in the format of \"projects/_/instances/{firebaseioSubdomain}/refs/{ref=**}\"");
	}
	const [, project, dbInstanceName, path] = match;
	if (project !== "_") {
		throw new Error(`Expect project to be '_' in a Firebase Realtime Database event`);
	}
	const emuHost = process.env.FIREBASE_DATABASE_EMULATOR_HOST;
	if (emuHost) {
		const dbInstance = `http://${emuHost}/?ns=${dbInstanceName}`;
		return [dbInstance, path];
	} else {
		const dbInstance = "https://" + dbInstanceName + "." + domain;
		return [dbInstance, path];
	}
}


exports.DataSnapshot = require_common_providers_database.DataSnapshot;
exports.InstanceBuilder = InstanceBuilder;
exports.RefBuilder = RefBuilder;
exports._instanceWithOptions = _instanceWithOptions;
exports._refWithOptions = _refWithOptions;
Object.defineProperty(exports, 'database_exports', {
  enumerable: true,
  get: function () {
    return database_exports;
  }
});
exports.extractInstanceAndPath = extractInstanceAndPath;
exports.instance = instance;
exports.provider = provider;
exports.ref = ref;
exports.service = service;
