// firebase-functions/lib/v1/providers/pubsub.js
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.js');
const require_v1_cloud_functions = require('../cloud-functions.js');


var pubsub_exports =  require_rolldown_runtime.__export({
	Message: () => Message,
	ScheduleBuilder: () => ScheduleBuilder,
	TopicBuilder: () => TopicBuilder,
	_scheduleWithOptions: () => _scheduleWithOptions,
	_topicWithOptions: () => _topicWithOptions,
	provider: () => provider,
	schedule: () => schedule,
	service: () => service,
	topic: () => topic
});

const provider = "google.pubsub";

const service = "pubsub.googleapis.com";

function topic(topic$1) {
	return _topicWithOptions(topic$1, {});
}

function _topicWithOptions(topic$1, options) {
	if (topic$1.indexOf("/") !== -1) {
		throw new Error("Topic name may not have a /");
	}
	return new TopicBuilder(() => {
		if (!process.env.GCLOUD_PROJECT) {
			throw new Error("process.env.GCLOUD_PROJECT is not set.");
		}
		return `projects/${process.env.GCLOUD_PROJECT}/topics/${topic$1}`;
	}, options);
}

var TopicBuilder = class {
	
	constructor(triggerResource, options) {
		this.triggerResource = triggerResource;
		this.options = options;
	}
	
	onPublish(handler) {
		return require_v1_cloud_functions.makeCloudFunction({
			handler,
			provider,
			service,
			triggerResource: this.triggerResource,
			eventType: "topic.publish",
			dataConstructor: (raw) => new Message(raw.data),
			options: this.options
		});
	}
};

function schedule(schedule$1) {
	return _scheduleWithOptions(schedule$1, {});
}

function _scheduleWithOptions(schedule$1, options) {
	const triggerResource = () => {
		if (!process.env.GCLOUD_PROJECT) {
			throw new Error("process.env.GCLOUD_PROJECT is not set.");
		}
		return `projects/${process.env.GCLOUD_PROJECT}/topics`;
	};
	return new ScheduleBuilder(triggerResource, {
		...options,
		schedule: { schedule: schedule$1 }
	});
}

var ScheduleBuilder = class {
	
	constructor(triggerResource, options) {
		this.triggerResource = triggerResource;
		this.options = options;
	}
	retryConfig(config) {
		this.options.schedule.retryConfig = config;
		return this;
	}
	timeZone(timeZone) {
		this.options.schedule.timeZone = timeZone;
		return this;
	}
	
	onRun(handler) {
		const cloudFunction = require_v1_cloud_functions.makeCloudFunction({
			contextOnlyHandler: handler,
			provider,
			service,
			triggerResource: this.triggerResource,
			eventType: "topic.publish",
			options: this.options,
			labels: { "deployment-scheduled": "true" }
		});
		return cloudFunction;
	}
};

var Message = class {
	constructor(data) {
		[this.data, this.attributes, this._json] = [
			data.data,
			data.attributes || {},
			data.json
		];
	}
	
	get json() {
		if (typeof this._json === "undefined") {
			this._json = JSON.parse(Buffer.from(this.data, "base64").toString("utf8"));
		}
		return this._json;
	}
	
	toJSON() {
		return {
			data: this.data,
			attributes: this.attributes
		};
	}
};


exports.Message = Message;
exports.ScheduleBuilder = ScheduleBuilder;
exports.TopicBuilder = TopicBuilder;
exports._scheduleWithOptions = _scheduleWithOptions;
exports._topicWithOptions = _topicWithOptions;
exports.provider = provider;
Object.defineProperty(exports, 'pubsub_exports', {
  enumerable: true,
  get: function () {
    return pubsub_exports;
  }
});
exports.schedule = schedule;
exports.service = service;
exports.topic = topic;
