// firebase-functions/lib/v1/providers/remoteConfig.js
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.js');
const require_v1_cloud_functions = require('../cloud-functions.js');


var remoteConfig_exports =  require_rolldown_runtime.__export({
	UpdateBuilder: () => UpdateBuilder,
	_onUpdateWithOptions: () => _onUpdateWithOptions,
	onUpdate: () => onUpdate,
	provider: () => provider,
	service: () => service
});

const provider = "google.firebase.remoteconfig";

const service = "firebaseremoteconfig.googleapis.com";

function onUpdate(handler) {
	return _onUpdateWithOptions(handler, {});
}

function _onUpdateWithOptions(handler, options) {
	const triggerResource = () => {
		if (!process.env.GCLOUD_PROJECT) {
			throw new Error("process.env.GCLOUD_PROJECT is not set.");
		}
		return `projects/${process.env.GCLOUD_PROJECT}`;
	};
	return new UpdateBuilder(triggerResource, options).onUpdate(handler);
}

var UpdateBuilder = class {
	
	constructor(triggerResource, options) {
		this.triggerResource = triggerResource;
		this.options = options;
	}
	
	onUpdate(handler) {
		return require_v1_cloud_functions.makeCloudFunction({
			handler,
			provider,
			service,
			triggerResource: this.triggerResource,
			eventType: "update",
			options: this.options
		});
	}
};


exports.UpdateBuilder = UpdateBuilder;
exports._onUpdateWithOptions = _onUpdateWithOptions;
exports.onUpdate = onUpdate;
exports.provider = provider;
Object.defineProperty(exports, 'remoteConfig_exports', {
  enumerable: true,
  get: function () {
    return remoteConfig_exports;
  }
});
exports.service = service;
