// firebase-functions/lib/v1/providers/storage.js
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.js');
const require_common_config = require('../../common/config.js');
const require_v1_cloud_functions = require('../cloud-functions.js');


var storage_exports =  require_rolldown_runtime.__export({
	BucketBuilder: () => BucketBuilder,
	ObjectBuilder: () => ObjectBuilder,
	_bucketWithOptions: () => _bucketWithOptions,
	_objectWithOptions: () => _objectWithOptions,
	bucket: () => bucket,
	object: () => object,
	provider: () => provider,
	service: () => service
});

const provider = "google.storage";

const service = "storage.googleapis.com";

function bucket(bucket$1) {
	return _bucketWithOptions({}, bucket$1);
}

function object() {
	return _objectWithOptions({});
}

function _bucketWithOptions(options, bucket$1) {
	const resourceGetter = () => {
		bucket$1 = bucket$1 || require_common_config.firebaseConfig().storageBucket;
		if (!bucket$1) {
			throw new Error("Missing bucket name. If you are unit testing, please provide a bucket name" + " through `functions.storage.bucket(bucketName)`, or set process.env.FIREBASE_CONFIG.");
		}
		if (!/^[a-z\d][a-z\d\\._-]{1,230}[a-z\d]$/.test(bucket$1)) {
			throw new Error(`Invalid bucket name ${bucket$1}`);
		}
		return `projects/_/buckets/${bucket$1}`;
	};
	return new BucketBuilder(resourceGetter, options);
}

function _objectWithOptions(options) {
	return _bucketWithOptions(options).object();
}

var BucketBuilder = class {
	
	constructor(triggerResource, options) {
		this.triggerResource = triggerResource;
		this.options = options;
	}
	
	object() {
		return new ObjectBuilder(this.triggerResource, this.options);
	}
};

var ObjectBuilder = class {
	
	constructor(triggerResource, options) {
		this.triggerResource = triggerResource;
		this.options = options;
	}
	
	onArchive(handler) {
		return this.onOperation(handler, "object.archive");
	}
	
	onDelete(handler) {
		return this.onOperation(handler, "object.delete");
	}
	
	onFinalize(handler) {
		return this.onOperation(handler, "object.finalize");
	}
	
	onMetadataUpdate(handler) {
		return this.onOperation(handler, "object.metadataUpdate");
	}
	
	onOperation(handler, eventType) {
		return require_v1_cloud_functions.makeCloudFunction({
			handler,
			provider,
			service,
			eventType,
			triggerResource: this.triggerResource,
			options: this.options
		});
	}
};


exports.BucketBuilder = BucketBuilder;
exports.ObjectBuilder = ObjectBuilder;
exports._bucketWithOptions = _bucketWithOptions;
exports._objectWithOptions = _objectWithOptions;
exports.bucket = bucket;
exports.object = object;
exports.provider = provider;
exports.service = service;
Object.defineProperty(exports, 'storage_exports', {
  enumerable: true,
  get: function () {
    return storage_exports;
  }
});
