// @google-cloud/firestore/build/src/aggregate.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateField = exports.Aggregate = void 0;
const path_1 = require("./path");
const assert = require("assert");

class Aggregate {
    constructor(alias, aggregateType, fieldPath) {
        this.alias = alias;
        this.aggregateType = aggregateType;
        this.fieldPath = fieldPath;
    }
    
    toProto() {
        const proto = {};
        if (this.aggregateType === 'count') {
            proto.count = {};
        }
        else if (this.aggregateType === 'sum') {
            assert(this.fieldPath !== undefined, 'Missing field path for sum aggregation.');
            proto.sum = {
                field: {
                    fieldPath: path_1.FieldPath.fromArgument(this.fieldPath).formattedName,
                },
            };
        }
        else if (this.aggregateType === 'avg') {
            assert(this.fieldPath !== undefined, 'Missing field path for average aggregation.');
            proto.avg = {
                field: {
                    fieldPath: path_1.FieldPath.fromArgument(this.fieldPath).formattedName,
                },
            };
        }
        else {
            throw new Error(`Aggregate type ${this.aggregateType} unimplemented.`);
        }
        proto.alias = this.alias;
        return proto;
    }
}
exports.Aggregate = Aggregate;

class AggregateField {
    
    constructor(aggregateType, field) {
        this.aggregateType = aggregateType;
        
        this.type = 'AggregateField';
        this._field = field;
    }
    
    isEqual(other) {
        return (other instanceof AggregateField &&
            this.aggregateType === other.aggregateType &&
            ((this._field === undefined && other._field === undefined) ||
                (this._field !== undefined &&
                    other._field !== undefined &&
                    path_1.FieldPath.fromArgument(this._field).isEqual(path_1.FieldPath.fromArgument(other._field)))));
    }
    
    static count() {
        return new AggregateField('count');
    }
    
    static average(field) {
        return new AggregateField('avg', field);
    }
    
    static sum(field) {
        return new AggregateField('sum', field);
    }
}
exports.AggregateField = AggregateField;
