// @google-cloud/firestore/build/src/convert.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.timestampFromJson = timestampFromJson;
exports.detectValueType = detectValueType;
exports.detectGoogleProtobufValueType = detectGoogleProtobufValueType;
exports.valueFromJson = valueFromJson;
exports.fieldsFromJson = fieldsFromJson;
const validate_1 = require("./validate");
const map_type_1 = require("./map-type");


function timestampFromJson(timestampValue, argumentName) {
    let timestampProto = {};
    if (typeof timestampValue === 'string') {
        const date = new Date(timestampValue);
        const seconds = Math.floor(date.getTime() / 1000);
        let nanos = 0;
        if (timestampValue.length > 20) {
            const nanoString = timestampValue.substring(20, timestampValue.length - 1);
            const trailingZeroes = 9 - nanoString.length;
            nanos = Number(nanoString) * Math.pow(10, trailingZeroes);
        }
        if (isNaN(seconds) || isNaN(nanos)) {
            argumentName = argumentName || 'timestampValue';
            throw new Error(`Specify a valid ISO 8601 timestamp for "${argumentName}".`);
        }
        timestampProto = {
            seconds: seconds || undefined,
            nanos: nanos || undefined,
        };
    }
    else if (timestampValue !== undefined) {
        (0, validate_1.validateObject)('timestampValue', timestampValue);
        timestampProto = {
            seconds: timestampValue.seconds || undefined,
            nanos: timestampValue.nanos || undefined,
        };
    }
    return timestampProto;
}

function bytesFromJson(bytesValue) {
    if (typeof bytesValue === 'string') {
        return Buffer.from(bytesValue, 'base64');
    }
    else {
        return bytesValue;
    }
}

function detectValueType(proto) {
    var _a;
    let valueType;
    if (proto.valueType) {
        valueType = proto.valueType;
    }
    else {
        const detectedValues = [];
        if (proto.stringValue !== undefined) {
            detectedValues.push('stringValue');
        }
        if (proto.booleanValue !== undefined) {
            detectedValues.push('booleanValue');
        }
        if (proto.integerValue !== undefined) {
            detectedValues.push('integerValue');
        }
        if (proto.doubleValue !== undefined) {
            detectedValues.push('doubleValue');
        }
        if (proto.timestampValue !== undefined) {
            detectedValues.push('timestampValue');
        }
        if (proto.referenceValue !== undefined) {
            detectedValues.push('referenceValue');
        }
        if (proto.arrayValue !== undefined) {
            detectedValues.push('arrayValue');
        }
        if (proto.nullValue !== undefined) {
            detectedValues.push('nullValue');
        }
        if (proto.mapValue !== undefined) {
            detectedValues.push('mapValue');
        }
        if (proto.geoPointValue !== undefined) {
            detectedValues.push('geoPointValue');
        }
        if (proto.bytesValue !== undefined) {
            detectedValues.push('bytesValue');
        }
        if (detectedValues.length !== 1) {
            throw new Error(`Unable to infer type value from '${JSON.stringify(proto)}'.`);
        }
        valueType = detectedValues[0];
    }

    if (valueType === 'mapValue') {
        const fields = (_a = proto.mapValue) === null || _a === void 0 ? void 0 : _a.fields;
        if (fields) {
            const props = Object.keys(fields);
            if (props.indexOf(map_type_1.RESERVED_MAP_KEY) !== -1 &&
                detectValueType(fields[map_type_1.RESERVED_MAP_KEY]) === 'stringValue' &&
                fields[map_type_1.RESERVED_MAP_KEY].stringValue === map_type_1.RESERVED_MAP_KEY_VECTOR_VALUE) {
                valueType = 'vectorValue';
            }
        }
    }
    return valueType;
}

function detectGoogleProtobufValueType(proto) {
    const detectedValues = [];
    if (proto.nullValue !== undefined) {
        detectedValues.push('nullValue');
    }
    if (proto.numberValue !== undefined) {
        detectedValues.push('numberValue');
    }
    if (proto.stringValue !== undefined) {
        detectedValues.push('stringValue');
    }
    if (proto.boolValue !== undefined) {
        detectedValues.push('boolValue');
    }
    if (proto.structValue !== undefined) {
        detectedValues.push('structValue');
    }
    if (proto.listValue !== undefined) {
        detectedValues.push('listValue');
    }
    if (detectedValues.length !== 1) {
        throw new Error(`Unable to infer type value from '${JSON.stringify(proto)}'.`);
    }
    return detectedValues[0];
}

function valueFromJson(fieldValue) {
    const valueType = detectValueType(fieldValue);
    switch (valueType) {
        case 'timestampValue':
            return {
                timestampValue: timestampFromJson(fieldValue.timestampValue),
            };
        case 'bytesValue':
            return {
                bytesValue: bytesFromJson(fieldValue.bytesValue),
            };
        case 'doubleValue':
            return {
                doubleValue: Number(fieldValue.doubleValue),
            };
        case 'arrayValue': {
            const arrayValue = [];
            if (Array.isArray(fieldValue.arrayValue.values)) {
                for (const value of fieldValue.arrayValue.values) {
                    arrayValue.push(valueFromJson(value));
                }
            }
            return {
                arrayValue: {
                    values: arrayValue,
                },
            };
        }
        case 'mapValue':
        case 'vectorValue': {
            const mapValue = {};
            const fields = fieldValue.mapValue.fields;
            if (fields) {
                for (const prop of Object.keys(fields)) {
                    mapValue[prop] = valueFromJson(fieldValue.mapValue.fields[prop]);
                }
            }
            return {
                mapValue: {
                    fields: mapValue,
                },
            };
        }
        default:
            return fieldValue;
    }
}

function fieldsFromJson(document) {
    const result = {};
    for (const prop of Object.keys(document)) {
        result[prop] = valueFromJson(document[prop]);
    }
    return result;
}
