// @google-cloud/firestore/build/src/filter.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.CompositeFilter = exports.UnaryFilter = exports.Filter = void 0;

class Filter {
    
    static where(fieldPath, opStr, value) {
        return new UnaryFilter(fieldPath, opStr, value);
    }
    
    static or(...filters) {
        return new CompositeFilter(filters, 'OR');
    }
    
    static and(...filters) {
        return new CompositeFilter(filters, 'AND');
    }
}
exports.Filter = Filter;

class UnaryFilter extends Filter {
    
    constructor(field, operator, value) {
        super();
        this.field = field;
        this.operator = operator;
        this.value = value;
    }
    
    _getField() {
        return this.field;
    }
    
    _getOperator() {
        return this.operator;
    }
    
    _getValue() {
        return this.value;
    }
}
exports.UnaryFilter = UnaryFilter;

class CompositeFilter extends Filter {
    
    constructor(filters, operator) {
        super();
        this.filters = filters;
        this.operator = operator;
    }
    
    _getFilters() {
        return this.filters;
    }
    
    _getOperator() {
        return this.operator;
    }
}
exports.CompositeFilter = CompositeFilter;
