// @google-cloud/firestore/build/src/geo-point.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.GeoPoint = void 0;
const validate_1 = require("./validate");

class GeoPoint {
    
    constructor(latitude, longitude) {
        (0, validate_1.validateNumber)('latitude', latitude, { minValue: -90, maxValue: 90 });
        (0, validate_1.validateNumber)('longitude', longitude, { minValue: -180, maxValue: 180 });
        this._latitude = latitude;
        this._longitude = longitude;
    }
    
    get latitude() {
        return this._latitude;
    }
    
    get longitude() {
        return this._longitude;
    }
    
    isEqual(other) {
        return (this === other ||
            (other instanceof GeoPoint &&
                this.latitude === other.latitude &&
                this.longitude === other.longitude));
    }
    
    toProto() {
        return {
            geoPointValue: {
                latitude: this.latitude,
                longitude: this.longitude,
            },
        };
    }
    
    static fromProto(proto) {
        return new GeoPoint(proto.latitude || 0, proto.longitude || 0);
    }
}
exports.GeoPoint = GeoPoint;
