// @google-cloud/firestore/build/src/pool.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientPool = exports.CLIENT_TERMINATED_ERROR_MSG = void 0;
const assert = require("assert");
const logger_1 = require("./logger");
const util_1 = require("./util");
exports.CLIENT_TERMINATED_ERROR_MSG = 'The client has already been terminated';

class ClientPool {
    
    constructor(concurrentOperationLimit, maxIdleClients, clientFactory, clientDestructor = () => Promise.resolve()) {
        this.concurrentOperationLimit = concurrentOperationLimit;
        this.maxIdleClients = maxIdleClients;
        this.clientFactory = clientFactory;
        this.clientDestructor = clientDestructor;
        this.grpcEnabled = false;
        
        this.activeClients = new Map();
        
        this.failedClients = new Set();
        
        this.clientIdByClient = new WeakMap();
        
        this.terminated = false;
        
        this.terminateDeferred = new util_1.Deferred();
        
        this.instanceId = 'cpl' + (0, util_1.requestTag)();
        this.lazyLogStringForAllClientIds = new LazyLogStringForAllClientIds({
            activeClients: this.activeClients,
            failedClients: this.failedClients,
            clientIdByClient: this.clientIdByClient,
        });
    }
    
    acquire(requestTag, requiresGrpc) {
        let selectedClient = null;
        let selectedClientRequestCount = -1;

        this.grpcEnabled = this.grpcEnabled || requiresGrpc;

        requiresGrpc = requiresGrpc || this.grpcEnabled;
        for (const [client, metadata] of this.activeClients) {



            if (!this.failedClients.has(client) &&
                metadata.activeRequestCount > selectedClientRequestCount &&
                metadata.activeRequestCount < this.concurrentOperationLimit &&
                (metadata.grpcEnabled || !requiresGrpc)) {
                selectedClient = client;
                selectedClientRequestCount = metadata.activeRequestCount;
            }
        }
        if (selectedClient) {
            const selectedClientId = this.clientIdByClient.get(selectedClient);
            (0, logger_1.logger)(`ClientPool[${this.instanceId}].acquire`, requestTag, 'Re-using existing client [%s] with %s remaining operations', selectedClientId, this.concurrentOperationLimit - selectedClientRequestCount);
        }
        else {
            const newClientId = 'cli' + (0, util_1.requestTag)();
            (0, logger_1.logger)(`ClientPool[${this.instanceId}].acquire`, requestTag, 'Creating a new client [%s] (requiresGrpc: %s)', newClientId, requiresGrpc);
            selectedClient = this.clientFactory(requiresGrpc);
            this.clientIdByClient.set(selectedClient, newClientId);
            selectedClientRequestCount = 0;
            assert(!this.activeClients.has(selectedClient), 'The provided client factory returned an existing instance');
        }
        this.activeClients.set(selectedClient, {
            grpcEnabled: requiresGrpc,
            activeRequestCount: selectedClientRequestCount + 1,
        });
        return selectedClient;
    }
    
    async release(requestTag, client) {
        const clientId = this.clientIdByClient.get(client);
        const metadata = this.activeClients.get(client);
        assert(metadata && metadata.activeRequestCount > 0, 'No active requests');
        this.activeClients.set(client, {
            grpcEnabled: metadata.grpcEnabled,
            activeRequestCount: metadata.activeRequestCount - 1,
        });
        if (this.terminated && this.opCount === 0) {
            this.terminateDeferred.resolve();
        }
        const gcDetermination = this.shouldGarbageCollectClient(client);
        (0, logger_1.logger)(`ClientPool[${this.instanceId}].release`, requestTag, 'Releasing client [%s] (gc=%s)', clientId, gcDetermination);
        if (!gcDetermination.shouldGarbageCollectClient) {
            return;
        }
        (0, logger_1.logger)(`ClientPool[${this.instanceId}].release`, requestTag, 'Garbage collecting client [%s] (%s)', clientId, this.lazyLogStringForAllClientIds);
        const activeClientDeleted = this.activeClients.delete(client);
        this.failedClients.delete(client);
        await this.clientDestructor(client);
        (0, logger_1.logger)(`ClientPool[${this.instanceId}].release`, requestTag, 'Garbage collected client [%s] activeClientDeleted=%s (%s)', clientId, activeClientDeleted, this.lazyLogStringForAllClientIds);
    }
    
    shouldGarbageCollectClient(client) {
        const clientMetadata = this.activeClients.get(client);
        if (clientMetadata.activeRequestCount !== 0) {

            return new ClientHasActiveRequests({
                shouldGarbageCollectClient: false,
                clientActiveRequestCount: clientMetadata.activeRequestCount,
            });
        }
        if (this.grpcEnabled !== clientMetadata.grpcEnabled) {

            return new PoolIsTransitioningToGrpc({
                shouldGarbageCollectClient: true,
                clientActiveRequestCount: clientMetadata.activeRequestCount,
                poolGrpcEnabled: this.grpcEnabled,
                clientGrpcEnabled: clientMetadata.grpcEnabled,
            });
        }


        if (this.failedClients.has(client)) {
            return new ClientIsFailed({
                shouldGarbageCollectClient: true,
                clientActiveRequestCount: clientMetadata.activeRequestCount,
            });
        }


        let idleCapacityCount = 0;
        for (const [, metadata] of this.activeClients) {
            idleCapacityCount +=
                this.concurrentOperationLimit - metadata.activeRequestCount;
        }
        const maxIdleCapacityCount = this.maxIdleClients * this.concurrentOperationLimit;
        return new IdleCapacity({
            shouldGarbageCollectClient: idleCapacityCount > maxIdleCapacityCount,
            clientActiveRequestCount: clientMetadata.activeRequestCount,
            idleCapacityCount: idleCapacityCount,
            maxIdleCapacityCount: maxIdleCapacityCount,
            maxIdleClients: this.maxIdleClients,
            concurrentOperationLimit: this.concurrentOperationLimit,
        });
    }
    

    get size() {
        return this.activeClients.size;
    }
    

    get opCount() {
        let activeOperationCount = 0;
        this.activeClients.forEach(metadata => (activeOperationCount += metadata.activeRequestCount));
        return activeOperationCount;
    }
    

    get _activeClients() {
        return this.activeClients;
    }
    
    run(requestTag, requiresGrpc, op) {
        if (this.terminated) {
            return Promise.reject(new Error(exports.CLIENT_TERMINATED_ERROR_MSG));
        }
        const client = this.acquire(requestTag, requiresGrpc);
        return op(client)
            .catch(async (err) => {
            var _a;
            if ((_a = err.message) === null || _a === void 0 ? void 0 : _a.match(/RST_STREAM/)) {



                this.failedClients.add(client);
            }
            await this.release(requestTag, client);
            return Promise.reject(err);
        })
            .then(async (res) => {
            await this.release(requestTag, client);
            return res;
        });
    }
    async terminate() {
        this.terminated = true;

        if (this.opCount > 0) {
            (0, logger_1.logger)(`ClientPool[${this.instanceId}].terminate`, 
             null, 'Waiting for %s pending operations to complete before terminating (%s)', this.opCount, this.lazyLogStringForAllClientIds);
            await this.terminateDeferred.promise;
        }
        (0, logger_1.logger)(`ClientPool[${this.instanceId}].terminate`, 
         null, 'Closing all active clients (%s)', this.lazyLogStringForAllClientIds);
        for (const [client] of this.activeClients) {
            this.activeClients.delete(client);
            await this.clientDestructor(client);
        }
    }
}
exports.ClientPool = ClientPool;

class LazyLogStringForAllClientIds {
    constructor(config) {
        this.activeClients = config.activeClients;
        this.failedClients = config.failedClients;
        this.clientIdByClient = config.clientIdByClient;
    }
    toString() {
        const activeClientsDescription = Array.from(this.activeClients.entries())
            .map(([client, metadata]) => `${this.clientIdByClient.get(client)}=${metadata.activeRequestCount}`)
            .sort()
            .join(', ');
        const failedClientsDescription = Array.from(this.failedClients)
            .map(client => `${this.clientIdByClient.get(client)}`)
            .sort()
            .join(', ');
        return (`${this.activeClients.size} active clients: {` +
            activeClientsDescription +
            '}, ' +
            `${this.failedClients.size} failed clients: {` +
            failedClientsDescription +
            '}');
    }
}

class BaseShouldGarbageCollectClientResult {
    
    toString() {
        const propertyStrings = [];
        for (const propertyName of Object.getOwnPropertyNames(this)) {
            const propertyValue = this[propertyName];
            propertyStrings.push(`${propertyName}=${propertyValue}`);
        }
        return '{' + propertyStrings.join(', ') + '}';
    }
}
class ClientHasActiveRequests extends BaseShouldGarbageCollectClientResult {
    constructor(args) {
        super();
        this.name = 'ClientHasActiveRequests';
        this.shouldGarbageCollectClient = args.shouldGarbageCollectClient;
        this.clientActiveRequestCount = args.clientActiveRequestCount;
    }
}
class PoolIsTransitioningToGrpc extends BaseShouldGarbageCollectClientResult {
    constructor(args) {
        super();
        this.name = 'PoolIsTransitioningToGrpc';
        this.shouldGarbageCollectClient = args.shouldGarbageCollectClient;
        this.clientActiveRequestCount = args.clientActiveRequestCount;
        this.poolGrpcEnabled = args.poolGrpcEnabled;
        this.clientGrpcEnabled = args.clientGrpcEnabled;
    }
}
class ClientIsFailed extends BaseShouldGarbageCollectClientResult {
    constructor(args) {
        super();
        this.name = 'ClientIsFailed';
        this.shouldGarbageCollectClient = args.shouldGarbageCollectClient;
        this.clientActiveRequestCount = args.clientActiveRequestCount;
    }
}
class IdleCapacity extends BaseShouldGarbageCollectClientResult {
    constructor(args) {
        super();
        this.name = 'IdleCapacity';
        this.shouldGarbageCollectClient = args.shouldGarbageCollectClient;
        this.clientActiveRequestCount = args.clientActiveRequestCount;
        this.idleCapacityCount = args.idleCapacityCount;
        this.maxIdleCapacityCount = args.maxIdleCapacityCount;
        this.maxIdleClients = args.maxIdleClients;
        this.concurrentOperationLimit = args.concurrentOperationLimit;
    }
}
