// @google-cloud/firestore/build/src/query-profile.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.ExplainResults = exports.ExplainMetrics = exports.ExecutionStats = exports.PlanSummary = void 0;

class PlanSummary {
    
    constructor(indexesUsed) {
        this.indexesUsed = indexesUsed;
    }
    
    static _fromProto(plan, serializer) {
        const indexes = [];
        if (plan && plan.indexesUsed) {
            for (const index of plan.indexesUsed) {
                indexes.push(serializer.decodeGoogleProtobufStruct(index));
            }
        }
        return new PlanSummary(indexes);
    }
}
exports.PlanSummary = PlanSummary;

class ExecutionStats {
    
    constructor(resultsReturned, executionDuration, readOperations, debugStats) {
        this.resultsReturned = resultsReturned;
        this.executionDuration = executionDuration;
        this.readOperations = readOperations;
        this.debugStats = debugStats;
    }
    
    static _fromProto(stats, serializer) {
        var _a, _b;
        if (stats) {
            return new ExecutionStats(Number(stats.resultsReturned), {
                seconds: Number((_a = stats.executionDuration) === null || _a === void 0 ? void 0 : _a.seconds),
                nanoseconds: Number((_b = stats.executionDuration) === null || _b === void 0 ? void 0 : _b.nanos),
            }, Number(stats.readOperations), serializer.decodeGoogleProtobufStruct(stats.debugStats));
        }
        return null;
    }
}
exports.ExecutionStats = ExecutionStats;

class ExplainMetrics {
    
    constructor(planSummary, executionStats) {
        this.planSummary = planSummary;
        this.executionStats = executionStats;
    }
    
    static _fromProto(metrics, serializer) {
        return new ExplainMetrics(PlanSummary._fromProto(metrics.planSummary, serializer), ExecutionStats._fromProto(metrics.executionStats, serializer));
    }
}
exports.ExplainMetrics = ExplainMetrics;

class ExplainResults {
    
    constructor(metrics, snapshot) {
        this.metrics = metrics;
        this.snapshot = snapshot;
    }
}
exports.ExplainResults = ExplainResults;
