// @google-cloud/firestore/build/src/timestamp.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.Timestamp = void 0;
const validate_1 = require("./validate");

const MS_TO_NANOS = 1000000;

const MIN_SECONDS = -62135596800;

const MAX_SECONDS = 253402300799;

class Timestamp {
    
    static now() {
        return Timestamp.fromMillis(Date.now());
    }
    
    static fromDate(date) {
        return Timestamp.fromMillis(date.getTime());
    }
    
    static fromMillis(milliseconds) {
        const seconds = Math.floor(milliseconds / 1000);
        const nanos = Math.floor((milliseconds - seconds * 1000) * MS_TO_NANOS);
        return new Timestamp(seconds, nanos);
    }
    
    static fromProto(timestamp) {
        return new Timestamp(Number(timestamp.seconds || 0), timestamp.nanos || 0);
    }
    
    constructor(seconds, nanoseconds) {
        (0, validate_1.validateInteger)('seconds', seconds, {
            minValue: MIN_SECONDS,
            maxValue: MAX_SECONDS,
        });
        (0, validate_1.validateInteger)('nanoseconds', nanoseconds, {
            minValue: 0,
            maxValue: 999999999,
        });
        this._seconds = seconds;
        this._nanoseconds = nanoseconds;
    }
    
    get seconds() {
        return this._seconds;
    }
    
    get nanoseconds() {
        return this._nanoseconds;
    }
    
    toDate() {
        return new Date(this._seconds * 1000 + Math.round(this._nanoseconds / MS_TO_NANOS));
    }
    
    toMillis() {
        return this._seconds * 1000 + Math.floor(this._nanoseconds / MS_TO_NANOS);
    }
    
    isEqual(other) {
        return (this === other ||
            (other instanceof Timestamp &&
                this._seconds === other.seconds &&
                this._nanoseconds === other.nanoseconds));
    }
    
    toProto() {
        const timestamp = {};
        if (this.seconds) {
            timestamp.seconds = this.seconds.toString();
        }
        if (this.nanoseconds) {
            timestamp.nanos = this.nanoseconds;
        }
        return { timestampValue: timestamp };
    }
    
    valueOf() {






        const adjustedSeconds = this.seconds - MIN_SECONDS;

        const formattedSeconds = String(adjustedSeconds).padStart(12, '0');
        const formattedNanoseconds = String(this.nanoseconds).padStart(9, '0');
        return formattedSeconds + '.' + formattedNanoseconds;
    }
}
exports.Timestamp = Timestamp;
