// @grpc/grpc-js/build/src/priority-queue.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.PriorityQueue = void 0;
const top = 0;
const parent = (i) => Math.floor(i / 2);
const left = (i) => i * 2 + 1;
const right = (i) => i * 2 + 2;

class PriorityQueue {
    
    constructor(comparator = (a, b) => a > b) {
        this.comparator = comparator;
        this.heap = [];
    }
    
    size() {
        return this.heap.length;
    }
    
    isEmpty() {
        return this.size() == 0;
    }
    
    peek() {
        return this.heap[top];
    }
    
    push(...values) {
        values.forEach(value => {
            this.heap.push(value);
            this.siftUp();
        });
        return this.size();
    }
    
    pop() {
        const poppedValue = this.peek();
        const bottom = this.size() - 1;
        if (bottom > top) {
            this.swap(top, bottom);
        }
        this.heap.pop();
        this.siftDown();
        return poppedValue;
    }
    
    replace(value) {
        const replacedValue = this.peek();
        this.heap[top] = value;
        this.siftDown();
        return replacedValue;
    }
    greater(i, j) {
        return this.comparator(this.heap[i], this.heap[j]);
    }
    swap(i, j) {
        [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
    }
    siftUp() {
        let node = this.size() - 1;
        while (node > top && this.greater(node, parent(node))) {
            this.swap(node, parent(node));
            node = parent(node);
        }
    }
    siftDown() {
        let node = top;
        while ((left(node) < this.size() && this.greater(left(node), node)) ||
            (right(node) < this.size() && this.greater(right(node), node))) {
            let maxChild = (right(node) < this.size() && this.greater(right(node), left(node))) ? right(node) : left(node);
            this.swap(node, maxChild);
            node = maxChild;
        }
    }
}
exports.PriorityQueue = PriorityQueue;
