// firebase-admin/lib/database/index.js

"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseDatabaseError = exports.ServerValue = exports.enableLogging = void 0;
exports.getDatabase = getDatabase;
exports.getDatabaseWithUrl = getDatabaseWithUrl;
const standalone_1 = require("@firebase/database-compat/standalone");
const app_1 = require("../app");
const database_1 = require("./database");


exports.enableLogging = standalone_1.enableLogging;


exports.ServerValue = standalone_1.ServerValue;

function getDatabase(app) {
    return getDatabaseInstance({ app });
}

function getDatabaseWithUrl(url, app) {
    return getDatabaseInstance({ url, app });
}
function getDatabaseInstance(options) {
    let { app } = options;
    if (typeof app === 'undefined') {
        app = (0, app_1.getApp)();
    }
    const firebaseApp = app;
    const dbService = firebaseApp.getOrInitService('database', (app) => new database_1.DatabaseService(app));
    return dbService.getDatabase(options.url);
}
var error_1 = require("../utils/error");
Object.defineProperty(exports, "FirebaseDatabaseError", { enumerable: true, get: function () { return error_1.FirebaseDatabaseError; } });
