// teeny-request/build/src/TeenyStatistics.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.TeenyStatistics = exports.TeenyStatisticsWarning = void 0;

class TeenyStatisticsWarning extends Error {
    
    constructor(message) {
        super(message);
        this.threshold = 0;
        this.type = '';
        this.value = 0;
        this.name = this.constructor.name;
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.TeenyStatisticsWarning = TeenyStatisticsWarning;
TeenyStatisticsWarning.CONCURRENT_REQUESTS = 'ConcurrentRequestsExceededWarning';

class TeenyStatistics {
    
    constructor(opts) {
        
        this._concurrentRequests = 0;
        
        this._didConcurrentRequestWarn = false;
        this._options = TeenyStatistics._prepareOptions(opts);
    }
    
    getOptions() {
        return Object.assign({}, this._options);
    }
    
    setOptions(opts) {
        const oldOpts = this._options;
        this._options = TeenyStatistics._prepareOptions(opts);
        return oldOpts;
    }
    
    get counters() {
        return {
            concurrentRequests: this._concurrentRequests,
        };
    }
    
    requestStarting() {
        this._concurrentRequests++;
        if (this._options.concurrentRequests > 0 &&
            this._concurrentRequests >= this._options.concurrentRequests &&
            !this._didConcurrentRequestWarn) {
            this._didConcurrentRequestWarn = true;
            const warning = new TeenyStatisticsWarning('Possible excessive concurrent requests detected. ' +
                this._concurrentRequests +
                ' requests in-flight, which exceeds the configured threshold of ' +
                this._options.concurrentRequests +
                '. Use the TEENY_REQUEST_WARN_CONCURRENT_REQUESTS environment ' +
                'variable or the concurrentRequests option of teeny-request to ' +
                'increase or disable (0) this warning.');
            warning.type = TeenyStatisticsWarning.CONCURRENT_REQUESTS;
            warning.value = this._concurrentRequests;
            warning.threshold = this._options.concurrentRequests;
            process.emitWarning(warning);
        }
    }
    
    requestFinished() {

        this._concurrentRequests--;
    }
    
    static _prepareOptions({ concurrentRequests: diConcurrentRequests, } = {}) {
        let concurrentRequests = this.DEFAULT_WARN_CONCURRENT_REQUESTS;
        const envConcurrentRequests = Number(process.env.TEENY_REQUEST_WARN_CONCURRENT_REQUESTS);
        if (diConcurrentRequests !== undefined) {
            concurrentRequests = diConcurrentRequests;
        }
        else if (!Number.isNaN(envConcurrentRequests)) {
            concurrentRequests = envConcurrentRequests;
        }
        return { concurrentRequests };
    }
}
exports.TeenyStatistics = TeenyStatistics;

TeenyStatistics.DEFAULT_WARN_CONCURRENT_REQUESTS = 5000;
