// google-gax/build/src/bundlingCalls/bundleApiCaller.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.BundleApiCaller = void 0;
const call_1 = require("../call");
const googleError_1 = require("../googleError");

class BundleApiCaller {
    constructor(bundler) {
        this.bundler = bundler;
    }
    init(callback) {
        if (callback) {
            return new call_1.OngoingCall(callback);
        }
        return new call_1.OngoingCallPromise();
    }
    wrap(func) {
        return func;
    }
    call(apiCall, argument, settings, status) {
        if (!settings.isBundling) {
            throw new googleError_1.GoogleError('Bundling enabled with no isBundling!');
        }
        status.call((argument, callback) => {
            this.bundler.schedule(apiCall, argument, callback);
            return status;
        }, argument);
    }
    fail(canceller, err) {
        canceller.callback(err);
    }
    result(canceller) {
        return canceller.promise;
    }
}
exports.BundleApiCaller = BundleApiCaller;
