// firebase-admin/lib/remote-config/remote-config-api.js

"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomSignalOperator = exports.PercentConditionOperator = void 0;

var PercentConditionOperator;
(function (PercentConditionOperator) {
    
    PercentConditionOperator["UNKNOWN"] = "UNKNOWN";
    
    PercentConditionOperator["LESS_OR_EQUAL"] = "LESS_OR_EQUAL";
    
    PercentConditionOperator["GREATER_THAN"] = "GREATER_THAN";
    
    PercentConditionOperator["BETWEEN"] = "BETWEEN";
})(PercentConditionOperator || (exports.PercentConditionOperator = PercentConditionOperator = {}));

var CustomSignalOperator;
(function (CustomSignalOperator) {
    
    CustomSignalOperator["UNKNOWN"] = "UNKNOWN";
    
    CustomSignalOperator["NUMERIC_LESS_THAN"] = "NUMERIC_LESS_THAN";
    
    CustomSignalOperator["NUMERIC_LESS_EQUAL"] = "NUMERIC_LESS_EQUAL";
    
    CustomSignalOperator["NUMERIC_EQUAL"] = "NUMERIC_EQUAL";
    
    CustomSignalOperator["NUMERIC_NOT_EQUAL"] = "NUMERIC_NOT_EQUAL";
    
    CustomSignalOperator["NUMERIC_GREATER_THAN"] = "NUMERIC_GREATER_THAN";
    
    CustomSignalOperator["NUMERIC_GREATER_EQUAL"] = "NUMERIC_GREATER_EQUAL";
    
    CustomSignalOperator["STRING_CONTAINS"] = "STRING_CONTAINS";
    
    CustomSignalOperator["STRING_DOES_NOT_CONTAIN"] = "STRING_DOES_NOT_CONTAIN";
    
    CustomSignalOperator["STRING_EXACTLY_MATCHES"] = "STRING_EXACTLY_MATCHES";
    
    CustomSignalOperator["STRING_CONTAINS_REGEX"] = "STRING_CONTAINS_REGEX";
    
    CustomSignalOperator["SEMANTIC_VERSION_LESS_THAN"] = "SEMANTIC_VERSION_LESS_THAN";
    
    CustomSignalOperator["SEMANTIC_VERSION_LESS_EQUAL"] = "SEMANTIC_VERSION_LESS_EQUAL";
    
    CustomSignalOperator["SEMANTIC_VERSION_EQUAL"] = "SEMANTIC_VERSION_EQUAL";
    
    CustomSignalOperator["SEMANTIC_VERSION_NOT_EQUAL"] = "SEMANTIC_VERSION_NOT_EQUAL";
    
    CustomSignalOperator["SEMANTIC_VERSION_GREATER_THAN"] = "SEMANTIC_VERSION_GREATER_THAN";
    
    CustomSignalOperator["SEMANTIC_VERSION_GREATER_EQUAL"] = "SEMANTIC_VERSION_GREATER_EQUAL";
})(CustomSignalOperator || (exports.CustomSignalOperator = CustomSignalOperator = {}));
