// protobufjs/src/message.js
"use strict";
module.exports = Message;

var util = require("./util/minimal");


function Message(properties) {

    if (properties)
        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
            this[keys[i]] = properties[keys[i]];
}








Message.create = function create(properties) {
    return this.$type.create(properties);
};


Message.encode = function encode(message, writer) {
    return this.$type.encode(message, writer);
};


Message.encodeDelimited = function encodeDelimited(message, writer) {
    return this.$type.encodeDelimited(message, writer);
};


Message.decode = function decode(reader) {
    return this.$type.decode(reader);
};


Message.decodeDelimited = function decodeDelimited(reader) {
    return this.$type.decodeDelimited(reader);
};


Message.verify = function verify(message) {
    return this.$type.verify(message);
};


Message.fromObject = function fromObject(object) {
    return this.$type.fromObject(object);
};


Message.toObject = function toObject(message, options) {
    return this.$type.toObject(message, options);
};


Message.prototype.toJSON = function toJSON() {
    return this.$type.toObject(this, util.toJSONOptions);
};
