// @google-cloud/storage/build/esm/src/util.js













import * as path from 'path';
import * as querystring from 'querystring';
import { PassThrough } from 'stream';
import * as url from 'url';


import { getPackageJSON } from './package-json-helper.cjs';

const fileURLToPath = url.fileURLToPath;
const isEsm = true;
export function normalize(optionsOrCallback, cb) {
    const options = (typeof optionsOrCallback === 'object' ? optionsOrCallback : {});
    const callback = (typeof optionsOrCallback === 'function' ? optionsOrCallback : cb);
    return { options, callback };
}

export function objectEntries(obj) {
    return Object.keys(obj).map(key => [key, obj[key]]);
}

export function fixedEncodeURIComponent(str) {
    return encodeURIComponent(str).replace(/[!'()*]/g, c => '%' + c.charCodeAt(0).toString(16).toUpperCase());
}

export function encodeURI(uri, encodeSlash) {


    return uri
        .split('/')
        .map(fixedEncodeURIComponent)
        .join(encodeSlash ? '%2F' : '/');
}

export function qsStringify(qs) {
    return querystring.stringify(qs, '&', '=', {
        encodeURIComponent: (component) => encodeURI(component, true),
    });
}
export function objectKeyToLowercase(object) {
    const newObj = {};
    for (let key of Object.keys(object)) {
        const value = object[key];
        key = key.toLowerCase();
        newObj[key] = value;
    }
    return newObj;
}

export function unicodeJSONStringify(obj) {
    return JSON.stringify(obj).replace(/[\u0080-\uFFFF]/g, (char) => '\\u' + ('0000' + char.charCodeAt(0).toString(16)).slice(-4));
}

export function convertObjKeysToSnakeCase(obj) {
    if (obj instanceof Date || obj instanceof RegExp) {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.map(convertObjKeysToSnakeCase);
    }
    if (obj instanceof Object) {
        return Object.keys(obj).reduce((acc, cur) => {
            const s = cur[0].toLocaleLowerCase() +
                cur.slice(1).replace(/([A-Z]+)/g, (match, p1) => {
                    return `_${p1.toLowerCase()}`;
                });
            acc[s] = convertObjKeysToSnakeCase(obj[cur]);
            return acc;
        }, Object());
    }
    return obj;
}

export function formatAsUTCISO(dateTimeToFormat, includeTime = false, dateDelimiter = '', timeDelimiter = '') {
    const year = dateTimeToFormat.getUTCFullYear();
    const month = dateTimeToFormat.getUTCMonth() + 1;
    const day = dateTimeToFormat.getUTCDate();
    const hour = dateTimeToFormat.getUTCHours();
    const minute = dateTimeToFormat.getUTCMinutes();
    const second = dateTimeToFormat.getUTCSeconds();
    let resultString = `${year.toString().padStart(4, '0')}${dateDelimiter}${month
        .toString()
        .padStart(2, '0')}${dateDelimiter}${day.toString().padStart(2, '0')}`;
    if (includeTime) {
        resultString = `${resultString}T${hour
            .toString()
            .padStart(2, '0')}${timeDelimiter}${minute
            .toString()
            .padStart(2, '0')}${timeDelimiter}${second.toString().padStart(2, '0')}Z`;
    }
    return resultString;
}

export function getRuntimeTrackingString() {
    if (


    globalThis.Deno &&


        globalThis.Deno.version &&


        globalThis.Deno.version.deno) {


        return `gl-deno/${globalThis.Deno.version.deno}`;
    }
    else {
        return `gl-node/${process.versions.node}`;
    }
}

export function getUserAgentString() {
    const pkg = getPackageJSON();
    const hyphenatedPackageName = pkg.name
        .replace('@google-cloud', 'gcloud-node') // For legacy purposes.
        .replace('/', '-'); // For UA spec-compliance purposes.
    return hyphenatedPackageName + '/' + pkg.version;
}
export function getDirName() {
    let dirToUse = '';
    try {
        dirToUse = __dirname;
    }
    catch (e) {


        dirToUse = path.dirname(fileURLToPath(import.meta.url));
    }
    return dirToUse;
}
export function getModuleFormat() {
    return isEsm ? 'ESM' : 'CJS';
}
export class PassThroughShim extends PassThrough {
    constructor() {
        super(...arguments);
        this.shouldEmitReading = true;
        this.shouldEmitWriting = true;
    }
    _read(size) {
        if (this.shouldEmitReading) {
            this.emit('reading');
            this.shouldEmitReading = false;
        }
        super._read(size);
    }
    _write(chunk, encoding, callback) {
        if (this.shouldEmitWriting) {
            this.emit('writing');
            this.shouldEmitWriting = false;
        }

        process.nextTick(() => {
            super._write(chunk, encoding, callback);
        });
    }
    _final(callback) {


        if (this.shouldEmitReading) {
            this.emit('reading');
            this.shouldEmitReading = false;
        }
        if (this.shouldEmitWriting) {
            this.emit('writing');
            this.shouldEmitWriting = false;
        }
        callback(null);
    }
}
