// google-auth-library/build/src/transporters.js
"use strict";













Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultTransporter = void 0;
const gaxios_1 = require("gaxios");
const options_1 = require("./options");

const pkg = require('../../package.json');
const PRODUCT_NAME = 'google-api-nodejs-client';
class DefaultTransporter {
    constructor() {
        
        this.instance = new gaxios_1.Gaxios();
    }
    
    configure(opts = {}) {
        opts.headers = opts.headers || {};
        if (typeof window === 'undefined') {

            const uaValue = opts.headers['User-Agent'];
            if (!uaValue) {
                opts.headers['User-Agent'] = DefaultTransporter.USER_AGENT;
            }
            else if (!uaValue.includes(`${PRODUCT_NAME}/`)) {
                opts.headers['User-Agent'] =
                    `${uaValue} ${DefaultTransporter.USER_AGENT}`;
            }

            if (!opts.headers['x-goog-api-client']) {
                const nodeVersion = process.version.replace(/^v/, '');
                opts.headers['x-goog-api-client'] = `gl-node/${nodeVersion}`;
            }
        }
        return opts;
    }
    
    request(opts) {

        opts = this.configure(opts);
        (0, options_1.validate)(opts);
        return this.instance.request(opts).catch(e => {
            throw this.processError(e);
        });
    }
    get defaults() {
        return this.instance.defaults;
    }
    set defaults(opts) {
        this.instance.defaults = opts;
    }
    
    processError(e) {
        const res = e.response;
        const err = e;
        const body = res ? res.data : null;
        if (res && body && body.error && res.status !== 200) {
            if (typeof body.error === 'string') {
                err.message = body.error;
                err.status = res.status;
            }
            else if (Array.isArray(body.error.errors)) {
                err.message = body.error.errors
                    .map((err2) => err2.message)
                    .join('\n');
                err.code = body.error.code;
                err.errors = body.error.errors;
            }
            else {
                err.message = body.error.message;
                err.code = body.error.code;
            }
        }
        else if (res && res.status >= 400) {

            err.message = body;
            err.status = res.status;
        }
        return err;
    }
}
exports.DefaultTransporter = DefaultTransporter;

DefaultTransporter.USER_AGENT = `${PRODUCT_NAME}/${pkg.version}`;
