// google-gax/build/src/paginationCalls/resourceCollector.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceCollector = void 0;

class ResourceCollector {
    constructor(apiCall, maxResults = -1) {
        this.apiCall = apiCall;
        this.resources = [];
        this.maxResults = maxResults;
    }
    callback(err, resources, nextPageRequest) {
        if (err) {

            this.rejectCallback(err);
            return;
        }

        for (const resource of resources) {
            this.resources.push(resource);
            if (this.resources.length === this.maxResults) {
                nextPageRequest = null;
                break;
            }
        }

        if (!nextPageRequest) {
            this.resolveCallback(this.resources);
            return;
        }

        const callback = (...args) => this.callback(...args);
        setImmediate(this.apiCall, nextPageRequest, callback);
    }
    processAllPages(firstRequest) {
        return new Promise((resolve, reject) => {
            this.resolveCallback = resolve;
            this.rejectCallback = reject;

            const callback = (...args) => this.callback(...args);
            setImmediate(this.apiCall, firstRequest, callback);
        });
    }
}
exports.ResourceCollector = ResourceCollector;
