// google-gax/build/src/call.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.OngoingCallPromise = exports.OngoingCall = void 0;
const status_1 = require("./status");
const googleError_1 = require("./googleError");
class OngoingCall {
    
    constructor(callback) {
        this.callback = callback;
        this.completed = false;
    }
    
    cancel() {
        if (this.completed) {
            return;
        }
        this.completed = true;
        if (this.cancelFunc) {
            this.cancelFunc();
        }
        else {
            const error = new googleError_1.GoogleError('cancelled');
            error.code = status_1.Status.CANCELLED;
            this.callback(error);
        }
    }
    
    call(func, argument) {
        if (this.completed) {
            return;
        }
        const canceller = func(argument, (err, response, next, rawResponse) => {
            this.completed = true;
            setImmediate(this.callback, err, response, next, rawResponse);
        });
        if (canceller instanceof Promise) {
            canceller.catch(err => {
                setImmediate(this.callback, new googleError_1.GoogleError(err), null, null, null);
            });
        }
        this.cancelFunc = () => canceller.cancel();
    }
}
exports.OngoingCall = OngoingCall;
class OngoingCallPromise extends OngoingCall {
    
    constructor() {
        let resolveCallback;
        let rejectCallback;
        const callback = (err, response, next, rawResponse) => {
            if (err) {

                if (err.metadata) {
                    rejectCallback(googleError_1.GoogleError.parseGRPCStatusDetails(err));
                }
                else {
                    rejectCallback(err);
                }
            }
            else if (response !== undefined) {
                resolveCallback([response, next || null, rawResponse || null]);
            }
            else {
                throw new googleError_1.GoogleError('Neither error nor response are defined');
            }
        };
        const promise = new Promise((resolve, reject) => {
            resolveCallback = resolve;
            rejectCallback = reject;
        });
        super(callback);
        this.promise = promise;
        this.promise.cancel = () => {
            this.cancel();
        };
    }
}
exports.OngoingCallPromise = OngoingCallPromise;
