// firebase-functions/lib/logger/index.js
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.js');
const require_common_trace = require('../common/trace.js');
const require_logger_common = require('./common.js');
let util = require("util");
util = require_rolldown_runtime.__toESM(util);


var logger_exports =  require_rolldown_runtime.__export({
	debug: () => debug,
	error: () => error,
	info: () => info,
	log: () => log,
	logger: () => logger,
	warn: () => warn,
	write: () => write
});

function removeCircular(obj, refs = new Set()) {
	if (typeof obj !== "object" || !obj) {
		return obj;
	}
	if (obj.toJSON && typeof obj.toJSON === "function") {
		return obj.toJSON();
	}
	if (refs.has(obj)) {
		return "[Circular]";
	} else {
		refs.add(obj);
	}
	let returnObj;
	if (Array.isArray(obj)) {
		returnObj = new Array(obj.length);
	} else {
		returnObj = {};
	}
	for (const k in obj) {
		if (obj.hasOwnProperty(k)) {
			try {
				if (refs.has(obj[k])) {
					returnObj[k] = "[Circular]";
				} else {
					returnObj[k] = removeCircular(obj[k], refs);
				}
			} catch {
				returnObj[k] = "[Error - cannot serialize]";
			}
		} else {
			returnObj[k] = "[Error - defined in the prototype but missing in the object]";
		}
	}
	refs.delete(obj);
	return returnObj;
}

function write(entry) {
	const ctx = require_common_trace.traceContext.getStore();
	if (ctx?.traceId) {
		entry["logging.googleapis.com/trace"] = `projects/${process.env.GCLOUD_PROJECT}/traces/${ctx.traceId}`;
	}
	require_logger_common.UNPATCHED_CONSOLE[require_logger_common.CONSOLE_SEVERITY[entry.severity]](JSON.stringify(removeCircular(entry)));
}

function debug(...args) {
	write(entryFromArgs("DEBUG", args));
}

function log(...args) {
	write(entryFromArgs("INFO", args));
}

function info(...args) {
	write(entryFromArgs("INFO", args));
}

function warn(...args) {
	write(entryFromArgs("WARNING", args));
}

function error(...args) {
	write(entryFromArgs("ERROR", args));
}

function entryFromArgs(severity, args) {
	let entry = {};
	const lastArg = args[args.length - 1];
	if (lastArg && typeof lastArg === "object" && lastArg.constructor === Object) {
		entry = args.pop();
	}
	let message = (0, util.format)(...args);
	if (severity === "ERROR" && !args.find((arg) => arg instanceof Error)) {
		message = new Error(message).stack || message;
	}
	const out = {
		...entry,
		severity
	};
	if (message) {
		out.message = message;
	}
	return out;
}

const logger = {
	write,
	debug,
	log,
	info,
	warn,
	error
};


exports.debug = debug;
exports.error = error;
exports.info = info;
exports.log = log;
exports.logger = logger;
Object.defineProperty(exports, 'logger_exports', {
  enumerable: true,
  get: function () {
    return logger_exports;
  }
});
exports.warn = warn;
exports.write = write;
