// fast-xml-parser/src/v5/OutputBuilders/BaseOutputBuilder.js
class BaseOutputBuilder{
  constructor(){

  }

  addAttribute(name, value){
    if(this.options.onAttribute){

      const v = this.options.onAttribute(name, value, this.tagName);
      if(v) this.attributes[v.name] = v.value;
    }else{
      name = this.options.attributes.prefix + name + this.options.attributes.suffix;
      this.attributes[name] = this.parseValue(value, this.options.attributes.valueParsers);
    }
  }

  
    parseValue = function(val, valParsers){
      for (let i = 0; i < valParsers.length; i++) {
        let valParser = valParsers[i];
        if(typeof valParser === "string"){
          valParser = this.registeredParsers[valParser];
        }
        if(valParser){
          val = valParser.parse(val);
        }
      }
      return val;
    }

  
  _addChild(key, val){}

  
  addComment(text){
    if(this.options.nameFor.comment)
      this._addChild(this.options.nameFor.comment, text);
  }



  addCdata(text){
    if (this.options.nameFor.cdata) {
      this._addChild(this.options.nameFor.cdata, text);
    } else {
      this.addRawValue(text || "");
    }
  }

  addRawValue = text => this.addValue(text);

  addDeclaration(){
    if(!this.options.declaration){
    }else{
      this.addPi("?xml");
    }
    this.attributes = {}
  }
}

module.exports = BaseOutputBuilder;
