// fast-xml-parser/src/v5/OutputBuilders/ParserOptionsBuilder.js
const trimParser = require("../valueParsers/trim")
const booleanParser = require("../valueParsers/booleanParser")
const currencyParser = require("../valueParsers/currency")
const numberParser = require("../valueParsers/number")

const defaultOptions={
  nameFor:{
    text: "#text",
    comment: "",
    cdata: "",
  },


  piTag: false,
  declaration: false, //"?xml"
  tags: {
    valueParsers: [





    ]
  },
  attributes:{
    prefix: "@_",
    suffix: "",
    groupBy: "",
    
    valueParsers: [





    ]
  },
  dataType:{

  }
}


const withJoin = ["trim","join", "number","boolean","currency"]
const withoutJoin = ["trim", "number","boolean","currency"]

function buildOptions(options){

  const finalOptions = { ... defaultOptions};


  finalOptions.tags.valueParsers.push(...withJoin)
  if(!this.preserveOrder)
    finalOptions.tags.valueParsers.push(...withoutJoin);


  finalOptions.attributes.valueParsers.push(...withJoin)


  copyProperties(finalOptions,options);
  return finalOptions;
}

function copyProperties(target, source) {
  for (let key in source) {
    if (source.hasOwnProperty(key)) {
      if (typeof source[key] === 'object' && !Array.isArray(source[key])) {

        if (typeof target[key] === 'undefined') {
          target[key] = {};
        }
        copyProperties(target[key], source[key]);
      } else {

        target[key] = source[key];
      }
    }
  }
}

function registerCommonValueParsers(options){
  return {
    "trim": new trimParser(),

    "boolean": new booleanParser(),
    "number": new numberParser({
          hex: true,
          leadingZeros: true,
          eNotation: true
        }),
    "currency": new currencyParser(),

  }
}

module.exports = {
  buildOptions : buildOptions,
  registerCommonValueParsers: registerCommonValueParsers
}
