// firebase-functions/lib/common/providers/database.js
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.js');
const require_common_config = require('../config.js');
const require_common_utilities_path = require('../utilities/path.js');
let firebase_admin_database = require("firebase-admin/database");
firebase_admin_database = require_rolldown_runtime.__toESM(firebase_admin_database);



var DataSnapshot = class DataSnapshot {
	constructor(data, path, app, instance) {
		this.app = app;
		const config = require_common_config.firebaseConfig();
		if (instance) {
			this.instance = instance;
		} else if (app) {
			this.instance = app.options.databaseURL;
		} else if (config.databaseURL) {
			this.instance = config.databaseURL;
		} else if (process.env.GCLOUD_PROJECT) {
			this.instance = "https://" + process.env.GCLOUD_PROJECT + "-default-rtdb.firebaseio.com";
		}
		this._path = path;
		this._data = data;
	}
	
	get ref() {
		if (!this.app) {
			throw new Error("Please supply a Firebase app in the constructor for DataSnapshot" + " in order to use the .ref method.");
		}
		if (!this._ref) {
			let db;
			if (this.instance) {
				db = firebase_admin_database.getDatabaseWithUrl(this.instance, this.app);
			} else {
				db = firebase_admin_database.getDatabase(this.app);
			}
			this._ref = db.ref(this._fullPath());
		}
		return this._ref;
	}
	
	get key() {
		const segments = require_common_utilities_path.pathParts(this._fullPath());
		const last = segments[segments.length - 1];
		return !last || last === "" ? null : last;
	}
	
	val() {
		const parts = require_common_utilities_path.pathParts(this._childPath);
		let source = this._data;
		if (source === null) {
			return null;
		}
		if (parts.length) {
			for (const part of parts) {
				if (typeof source === "undefined" || source === null) {
					return null;
				}
				source = source[part];
			}
		}
		const node = source ?? null;
		return this._checkAndConvertToArray(node);
	}
	
	exportVal() {
		return this.val();
	}
	
	getPriority() {
		return 0;
	}
	
	exists() {
		const val = this.val();
		if (typeof val === "undefined" || val === null) {
			return false;
		}
		if (typeof val === "object" && Object.keys(val).length === 0) {
			return false;
		}
		return true;
	}
	
	child(childPath) {
		if (!childPath) {
			return this;
		}
		return this._dup(childPath);
	}
	
	forEach(action) {
		const val = this.val() || {};
		if (typeof val === "object") {
			return Object.keys(val).some((key) => action(this.child(key)) === true);
		}
		return false;
	}
	
	hasChild(childPath) {
		return this.child(childPath).exists();
	}
	
	hasChildren() {
		const val = this.val();
		return val !== null && typeof val === "object" && Object.keys(val).length > 0;
	}
	
	numChildren() {
		const val = this.val();
		return val !== null && typeof val === "object" ? Object.keys(val).length : 0;
	}
	
	toJSON() {
		return this.val();
	}
	
	_checkAndConvertToArray(node) {
		if (node === null || typeof node === "undefined") {
			return null;
		}
		if (typeof node !== "object") {
			return node;
		}
		const obj = {};
		let numKeys = 0;
		let maxKey = 0;
		let allIntegerKeys = true;
		for (const key in node) {
			if (!node.hasOwnProperty(key)) {
				continue;
			}
			const childNode = node[key];
			const v = this._checkAndConvertToArray(childNode);
			if (v === null) {
				continue;
			}
			obj[key] = v;
			numKeys++;
			const integerRegExp = /^(0|[1-9]\d*)$/;
			if (allIntegerKeys && integerRegExp.test(key)) {
				maxKey = Math.max(maxKey, Number(key));
			} else {
				allIntegerKeys = false;
			}
		}
		if (numKeys === 0) {
			return null;
		}
		if (allIntegerKeys && maxKey < 2 * numKeys) {
			const array = [];
			for (const key of Object.keys(obj)) {
				array[key] = obj[key];
			}
			return array;
		}
		return obj;
	}
	
	_dup(childPath) {
		const dup = new DataSnapshot(this._data, undefined, this.app, this.instance);
		[dup._path, dup._childPath] = [this._path, this._childPath];
		if (childPath) {
			dup._childPath = require_common_utilities_path.joinPath(dup._childPath, childPath);
		}
		return dup;
	}
	
	_fullPath() {
		return (this._path || "") + "/" + (this._childPath || "");
	}
};


exports.DataSnapshot = DataSnapshot;
