// @firebase/util/postinstall.js


const { writeFile, readFile } = require('node:fs/promises');
const { pathToFileURL } = require('node:url');
const { isAbsolute, join } = require('node:path');

const ENV_VARIABLE = 'FIREBASE_WEBAPP_CONFIG';

async function getPartialConfig() {
  const envVariable = process.env[ENV_VARIABLE]?.trim();

  if (!envVariable) {
    return undefined;
  }



  if (envVariable.startsWith('{"')) {
    try {
      return JSON.parse(envVariable);
    } catch (e) {
      console.warn(
        `JSON payload in \$${ENV_VARIABLE} could not be parsed, ignoring.\n`,
        e
      );
      return undefined;
    }
  }

  const fileURL = pathToFileURL(
    isAbsolute(envVariable) ? envVariable : join(process.cwd(), envVariable)
  );

  try {
    const fileContents = await readFile(fileURL, 'utf-8');
    return JSON.parse(fileContents);
  } catch (e) {
    console.warn(
      `Contents of "${envVariable}" could not be parsed, ignoring \$${ENV_VARIABLE}.\n`,
      e
    );
    return undefined;
  }
}

async function getFinalConfig(partialConfig) {
  if (!partialConfig) {
    return undefined;
  }


  if (process.env.X_GOOGLE_TARGET_PLATFORM === 'fah') {
    return partialConfig;
  }
  const projectId = partialConfig.projectId || '-';


  if (projectId.startsWith('demo-')) {
    return partialConfig;
  }
  const appId = partialConfig.appId;
  const apiKey = partialConfig.apiKey;
  if (!appId || !apiKey) {
    console.warn(
      `Unable to fetch Firebase config, appId and apiKey are required, ignoring \$${ENV_VARIABLE}.`
    );
    return undefined;
  }

  const url = `https://firebase.googleapis.com/v1alpha/projects/${projectId}/apps/${appId}/webConfig`;

  try {
    const response = await fetch(url, {
      headers: { 'x-goog-api-key': apiKey }
    });
    if (!response.ok) {
      console.warn(
        `Unable to fetch Firebase config, ignoring \$${ENV_VARIABLE}.`
      );
      console.warn(
        `${url} returned ${response.statusText} (${response.status})`
      );
      try {
        console.warn((await response.json()).error.message);
      } catch (e) {}
      return undefined;
    }
    const json = await response.json();
    return { ...json, apiKey };
  } catch (e) {
    console.warn(
      `Unable to fetch Firebase config, ignoring \$${ENV_VARIABLE}.\n`,
      e
    );
    return undefined;
  }
}

function handleUnexpectedError(e) {
  console.warn(
    `Unexpected error encountered in @firebase/util postinstall script, ignoring \$${ENV_VARIABLE}.`
  );
  console.warn(e);
  process.exit(0);
}

getPartialConfig()
  .catch(handleUnexpectedError)
  .then(getFinalConfig)
  .catch(handleUnexpectedError)
  .then(async finalConfig => {
    const defaults = finalConfig && {
      config: finalConfig,
      emulatorHosts: {
        firestore: process.env.FIRESTORE_EMULATOR_HOST,
        database: process.env.FIREBASE_DATABASE_EMULATOR_HOST,
        storage: process.env.FIREBASE_STORAGE_EMULATOR_HOST,
        auth: process.env.FIREBASE_AUTH_EMULATOR_HOST
      }
    };

    await Promise.all([
      writeFile(
        join(__dirname, 'dist', 'postinstall.js'),
        `'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.getDefaultsFromPostinstall = () => (${JSON.stringify(defaults)});`
      ),
      writeFile(
        join(__dirname, 'dist', 'postinstall.mjs'),
        `const getDefaultsFromPostinstall = () => (${JSON.stringify(defaults)});
export { getDefaultsFromPostinstall };`
      )
    ]);

    process.exit(0);
  })
  .catch(handleUnexpectedError);
