// @opentelemetry/api/build/esnext/api/context.js

import { NoopContextManager } from '../context/NoopContextManager';
import { getGlobal, registerGlobal, unregisterGlobal, } from '../internal/global-utils';
import { DiagAPI } from './diag';
const API_NAME = 'context';
const NOOP_CONTEXT_MANAGER = new NoopContextManager();

export class ContextAPI {
    
    constructor() { }
    
    static getInstance() {
        if (!this._instance) {
            this._instance = new ContextAPI();
        }
        return this._instance;
    }
    
    setGlobalContextManager(contextManager) {
        return registerGlobal(API_NAME, contextManager, DiagAPI.instance());
    }
    
    active() {
        return this._getContextManager().active();
    }
    
    with(context, fn, thisArg, ...args) {
        return this._getContextManager().with(context, fn, thisArg, ...args);
    }
    
    bind(context, target) {
        return this._getContextManager().bind(context, target);
    }
    _getContextManager() {
        return getGlobal(API_NAME) || NOOP_CONTEXT_MANAGER;
    }
    
    disable() {
        this._getContextManager().disable();
        unregisterGlobal(API_NAME, DiagAPI.instance());
    }
}
